{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "id": "xsjdhdm_kP0B"
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from IPython.display import display\n",
    "\n",
    "import scipy.stats as stats\n",
    "import pymc as pm\n",
    "import arviz as az\n",
    "\n",
    "sns.set_style(\"whitegrid\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 1000
    },
    "id": "0sllXH-RkQtj",
    "outputId": "b45019a6-16d4-48a9-d7bc-96e3f1edd466"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>x</th>\n",
       "      <th>y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>18.314043</td>\n",
       "      <td>12.567804</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>11.975118</td>\n",
       "      <td>6.350926</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>20.188111</td>\n",
       "      <td>11.388653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>15.787947</td>\n",
       "      <td>10.970399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9.641064</td>\n",
       "      <td>6.453270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>9.763732</td>\n",
       "      <td>5.259726</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>21.278046</td>\n",
       "      <td>13.825541</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>13.725025</td>\n",
       "      <td>8.986913</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>22.607104</td>\n",
       "      <td>14.651106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>19.547665</td>\n",
       "      <td>13.347023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>14.149328</td>\n",
       "      <td>8.466048</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>17.378791</td>\n",
       "      <td>10.439042</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>14.946248</td>\n",
       "      <td>10.636866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>18.461684</td>\n",
       "      <td>12.503519</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>17.331988</td>\n",
       "      <td>10.641989</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>15.898963</td>\n",
       "      <td>9.508089</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>17.577769</td>\n",
       "      <td>10.445704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>7.985551</td>\n",
       "      <td>5.354352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>17.543578</td>\n",
       "      <td>11.659115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>15.593922</td>\n",
       "      <td>10.024372</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>12.578849</td>\n",
       "      <td>8.941056</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>17.246452</td>\n",
       "      <td>10.776894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>17.010485</td>\n",
       "      <td>9.114727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>20.031624</td>\n",
       "      <td>14.268405</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>9.254325</td>\n",
       "      <td>5.921747</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>12.892437</td>\n",
       "      <td>8.752293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>3.527357</td>\n",
       "      <td>3.435328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>15.942716</td>\n",
       "      <td>9.605975</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>20.824604</td>\n",
       "      <td>13.760496</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>18.399652</td>\n",
       "      <td>10.778062</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>17.867275</td>\n",
       "      <td>10.733726</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>8.651755</td>\n",
       "      <td>6.523497</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>14.539603</td>\n",
       "      <td>10.834496</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>15.163298</td>\n",
       "      <td>9.187076</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>9.380957</td>\n",
       "      <td>6.692674</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>18.924330</td>\n",
       "      <td>12.193523</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>18.814970</td>\n",
       "      <td>11.079747</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>10.992649</td>\n",
       "      <td>8.928648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>19.930212</td>\n",
       "      <td>13.499736</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>17.645326</td>\n",
       "      <td>11.111084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>17.584860</td>\n",
       "      <td>10.742423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>15.255471</td>\n",
       "      <td>9.272911</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>14.142585</td>\n",
       "      <td>9.240165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>16.324873</td>\n",
       "      <td>9.893786</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>16.883676</td>\n",
       "      <td>11.708367</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>13.678907</td>\n",
       "      <td>8.130473</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>14.441869</td>\n",
       "      <td>10.134124</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            x          y\n",
       "0   18.314043  12.567804\n",
       "1   11.975118   6.350926\n",
       "2   20.188111  11.388653\n",
       "3   15.787947  10.970399\n",
       "4    9.641064   6.453270\n",
       "5    9.763732   5.259726\n",
       "6   21.278046  13.825541\n",
       "7   13.725025   8.986913\n",
       "8   22.607104  14.651106\n",
       "9   19.547665  13.347023\n",
       "10  14.149328   8.466048\n",
       "11  17.378791  10.439042\n",
       "12  14.946248  10.636866\n",
       "13  18.461684  12.503519\n",
       "14  17.331988  10.641989\n",
       "15  15.898963   9.508089\n",
       "16  17.577769  10.445704\n",
       "17   7.985551   5.354352\n",
       "18  17.543578  11.659115\n",
       "19  15.593922  10.024372\n",
       "20  12.578849   8.941056\n",
       "21  17.246452  10.776894\n",
       "22  17.010485   9.114727\n",
       "23  20.031624  14.268405\n",
       "24   9.254325   5.921747\n",
       "25  12.892437   8.752293\n",
       "26   3.527357   3.435328\n",
       "27  15.942716   9.605975\n",
       "28  20.824604  13.760496\n",
       "29  18.399652  10.778062\n",
       "30  17.867275  10.733726\n",
       "31   8.651755   6.523497\n",
       "32  14.539603  10.834496\n",
       "33  15.163298   9.187076\n",
       "34   9.380957   6.692674\n",
       "35  18.924330  12.193523\n",
       "36  18.814970  11.079747\n",
       "37  10.992649   8.928648\n",
       "38  19.930212  13.499736\n",
       "39  17.645326  11.111084\n",
       "40  17.584860  10.742423\n",
       "41  15.255471   9.272911\n",
       "42  14.142585   9.240165\n",
       "43  16.324873   9.893786\n",
       "44  16.883676  11.708367\n",
       "45  13.678907   8.130473\n",
       "46  14.441869  10.134124"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1 = pd.read_csv('data_01.csv')\n",
    "df2 = pd.read_csv('data_02.csv')\n",
    "df3 = pd.read_csv('data_03.csv')\n",
    "df4 = pd.read_csv('data_04.csv')\n",
    "df5 = pd.read_csv('data_05.csv')\n",
    "\n",
    "df5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "id": "QmrQJo6p-mWn"
   },
   "outputs": [],
   "source": [
    "df6 = pd.read_csv('data_06.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "w0wCfUOnynkf",
    "outputId": "c5748b88-3206-4a3d-9b69-df5a3eeb9671"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(55, 2)\n",
      "(50, 2)\n",
      "(55, 2)\n",
      "(53, 2)\n"
     ]
    }
   ],
   "source": [
    "# так как данные разной размерности, будем выкидывать выбросы из тех данных, в которых больше точек в 1,2,3,4 (методом подгонометрии :) )\n",
    "# q_low = df1.x.quantile(0.01)\n",
    "# q_hi  = df1.x.quantile(0.99)\n",
    "\n",
    "# df1 = df1[(df1.x < q_hi) & (df1.x > q_low)]\n",
    "# df1 = df1.reset_index(drop = True)\n",
    "print(df1.shape)\n",
    "####\n",
    "\n",
    "# q_low = df2.x.quantile(0.01)\n",
    "# q_hi  = df2.x.quantile(0.99)\n",
    "\n",
    "# df2 = df2[(df2.x < q_hi) & (df2.x > q_low)]\n",
    "# df2 = df2.reset_index(drop = True)\n",
    "print(df2.shape)\n",
    "\n",
    "####\n",
    "\n",
    "# q_low = df3.x.quantile(0.01)\n",
    "# q_hi  = df3.x.quantile(0.99)\n",
    "\n",
    "# df3 = df3[(df3.x < q_hi) & (df3.x > q_low)]\n",
    "# df3 = df3.reset_index(drop = True)\n",
    "print(df3.shape)\n",
    "###\n",
    "\n",
    "# q_low = df4.x.quantile(0.01)\n",
    "# q_hi  = df4.x.quantile(0.99)\n",
    "\n",
    "# df4 = df4[(df4.x < q_hi) & (df4.x > q_low)]\n",
    "# df4 = df4.reset_index(drop = True)\n",
    "print(df4.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "vXoK48RXz9y7",
    "outputId": "eaeab16e-e657-46c6-a5da-102d58f15d35"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(47, 2)\n"
     ]
    }
   ],
   "source": [
    "# q_low = df5.x.quantile(0.01)\n",
    "# q_hi  = df5.x.quantile(0.99)\n",
    "\n",
    "# df5 = df5[(df5.x < q_hi) & (df5.x > q_low)]\n",
    "# df = df5.reset_index(drop = True)x\n",
    "print(df5.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 423
    },
    "id": "wXWb1t4c1cy0",
    "outputId": "50101483-50cb-489e-abba-5aac31cfdfec"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>x</th>\n",
       "      <th>y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>13.344351</td>\n",
       "      <td>9.064982</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>12.141748</td>\n",
       "      <td>9.742935</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>16.045178</td>\n",
       "      <td>10.697097</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9.044173</td>\n",
       "      <td>7.119905</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>16.597604</td>\n",
       "      <td>11.721995</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>16.883676</td>\n",
       "      <td>11.708367</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>13.678907</td>\n",
       "      <td>8.130473</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>14.441869</td>\n",
       "      <td>10.134124</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>13.252132</td>\n",
       "      <td>8.762344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>12.047335</td>\n",
       "      <td>9.439947</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>259 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            x          y\n",
       "0   13.344351   9.064982\n",
       "1   12.141748   9.742935\n",
       "2   16.045178  10.697097\n",
       "3    9.044173   7.119905\n",
       "4   16.597604  11.721995\n",
       "..        ...        ...\n",
       "44  16.883676  11.708367\n",
       "45  13.678907   8.130473\n",
       "46  14.441869  10.134124\n",
       "0   13.252132   8.762344\n",
       "1   12.047335   9.439947\n",
       "\n",
       "[259 rows x 2 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.concat([df1, df2[0:47], df3, df4, df5, df6], axis=0).dropna() #довырезаем из 2 рандомную точку (да, плохо, но выдумывать ракету некогда)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 507
    },
    "id": "oUuUBptnkQ3a",
    "outputId": "d126c23c-cdd5-4670-c522-0841017e51f1"
   },
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "operands could not be broadcast together with shapes (237,) (259,) ",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[9], line 14\u001b[0m\n\u001b[1;32m     12\u001b[0m y_m \u001b[38;5;241m=\u001b[39m df\u001b[38;5;241m.\u001b[39my\n\u001b[1;32m     13\u001b[0m x_m \u001b[38;5;241m=\u001b[39m df\u001b[38;5;241m.\u001b[39mx\n\u001b[0;32m---> 14\u001b[0m y_m \u001b[38;5;241m=\u001b[39m alpha_real[idx] \u001b[38;5;241m+\u001b[39m beta_real[idx] \u001b[38;5;241m*\u001b[39m x_m \u001b[38;5;241m+\u001b[39m eps_real\n\u001b[1;32m     16\u001b[0m _, ax \u001b[38;5;241m=\u001b[39m plt\u001b[38;5;241m.\u001b[39msubplots(\u001b[38;5;241m2\u001b[39m, \u001b[38;5;241m3\u001b[39m, figsize\u001b[38;5;241m=\u001b[39m(\u001b[38;5;241m10\u001b[39m, \u001b[38;5;241m5\u001b[39m), sharex\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m, sharey\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[1;32m     17\u001b[0m ax \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mravel(ax)\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/pandas/core/generic.py:2102\u001b[0m, in \u001b[0;36mNDFrame.__array_ufunc__\u001b[0;34m(self, ufunc, method, *inputs, **kwargs)\u001b[0m\n\u001b[1;32m   2098\u001b[0m \u001b[38;5;129m@final\u001b[39m\n\u001b[1;32m   2099\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m__array_ufunc__\u001b[39m(\n\u001b[1;32m   2100\u001b[0m     \u001b[38;5;28mself\u001b[39m, ufunc: np\u001b[38;5;241m.\u001b[39mufunc, method: \u001b[38;5;28mstr\u001b[39m, \u001b[38;5;241m*\u001b[39minputs: Any, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs: Any\n\u001b[1;32m   2101\u001b[0m ):\n\u001b[0;32m-> 2102\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m arraylike\u001b[38;5;241m.\u001b[39marray_ufunc(\u001b[38;5;28mself\u001b[39m, ufunc, method, \u001b[38;5;241m*\u001b[39minputs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/pandas/core/arraylike.py:273\u001b[0m, in \u001b[0;36marray_ufunc\u001b[0;34m(self, ufunc, method, *inputs, **kwargs)\u001b[0m\n\u001b[1;32m    270\u001b[0m kwargs \u001b[38;5;241m=\u001b[39m _standardize_out_kwarg(\u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n\u001b[1;32m    272\u001b[0m \u001b[38;5;66;03m# for binary ops, use our custom dunder methods\u001b[39;00m\n\u001b[0;32m--> 273\u001b[0m result \u001b[38;5;241m=\u001b[39m maybe_dispatch_ufunc_to_dunder_op(\u001b[38;5;28mself\u001b[39m, ufunc, method, \u001b[38;5;241m*\u001b[39minputs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n\u001b[1;32m    274\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m result \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28mNotImplemented\u001b[39m:\n\u001b[1;32m    275\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m result\n",
      "File \u001b[0;32mops_dispatch.pyx:113\u001b[0m, in \u001b[0;36mpandas._libs.ops_dispatch.maybe_dispatch_ufunc_to_dunder_op\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/pandas/core/ops/common.py:76\u001b[0m, in \u001b[0;36m_unpack_zerodim_and_defer.<locals>.new_method\u001b[0;34m(self, other)\u001b[0m\n\u001b[1;32m     72\u001b[0m             \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mNotImplemented\u001b[39m\n\u001b[1;32m     74\u001b[0m other \u001b[38;5;241m=\u001b[39m item_from_zerodim(other)\n\u001b[0;32m---> 76\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m method(\u001b[38;5;28mself\u001b[39m, other)\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/pandas/core/arraylike.py:206\u001b[0m, in \u001b[0;36mOpsMixin.__rmul__\u001b[0;34m(self, other)\u001b[0m\n\u001b[1;32m    204\u001b[0m \u001b[38;5;129m@unpack_zerodim_and_defer\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m__rmul__\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    205\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m__rmul__\u001b[39m(\u001b[38;5;28mself\u001b[39m, other):\n\u001b[0;32m--> 206\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_arith_method(other, roperator\u001b[38;5;241m.\u001b[39mrmul)\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/pandas/core/series.py:5815\u001b[0m, in \u001b[0;36mSeries._arith_method\u001b[0;34m(self, other, op)\u001b[0m\n\u001b[1;32m   5813\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_arith_method\u001b[39m(\u001b[38;5;28mself\u001b[39m, other, op):\n\u001b[1;32m   5814\u001b[0m     \u001b[38;5;28mself\u001b[39m, other \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_align_for_op(other)\n\u001b[0;32m-> 5815\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m base\u001b[38;5;241m.\u001b[39mIndexOpsMixin\u001b[38;5;241m.\u001b[39m_arith_method(\u001b[38;5;28mself\u001b[39m, other, op)\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/pandas/core/base.py:1381\u001b[0m, in \u001b[0;36mIndexOpsMixin._arith_method\u001b[0;34m(self, other, op)\u001b[0m\n\u001b[1;32m   1378\u001b[0m     rvalues \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39marange(rvalues\u001b[38;5;241m.\u001b[39mstart, rvalues\u001b[38;5;241m.\u001b[39mstop, rvalues\u001b[38;5;241m.\u001b[39mstep)\n\u001b[1;32m   1380\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m np\u001b[38;5;241m.\u001b[39merrstate(\u001b[38;5;28mall\u001b[39m\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mignore\u001b[39m\u001b[38;5;124m\"\u001b[39m):\n\u001b[0;32m-> 1381\u001b[0m     result \u001b[38;5;241m=\u001b[39m ops\u001b[38;5;241m.\u001b[39marithmetic_op(lvalues, rvalues, op)\n\u001b[1;32m   1383\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_construct_result(result, name\u001b[38;5;241m=\u001b[39mres_name)\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/pandas/core/ops/array_ops.py:285\u001b[0m, in \u001b[0;36marithmetic_op\u001b[0;34m(left, right, op)\u001b[0m\n\u001b[1;32m    281\u001b[0m     _bool_arith_check(op, left, right)  \u001b[38;5;66;03m# type: ignore[arg-type]\u001b[39;00m\n\u001b[1;32m    283\u001b[0m     \u001b[38;5;66;03m# error: Argument 1 to \"_na_arithmetic_op\" has incompatible type\u001b[39;00m\n\u001b[1;32m    284\u001b[0m     \u001b[38;5;66;03m# \"Union[ExtensionArray, ndarray[Any, Any]]\"; expected \"ndarray[Any, Any]\"\u001b[39;00m\n\u001b[0;32m--> 285\u001b[0m     res_values \u001b[38;5;241m=\u001b[39m _na_arithmetic_op(left, right, op)  \u001b[38;5;66;03m# type: ignore[arg-type]\u001b[39;00m\n\u001b[1;32m    287\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m res_values\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/pandas/core/ops/array_ops.py:220\u001b[0m, in \u001b[0;36m_na_arithmetic_op\u001b[0;34m(left, right, op, is_cmp)\u001b[0m\n\u001b[1;32m    217\u001b[0m     func \u001b[38;5;241m=\u001b[39m partial(expressions\u001b[38;5;241m.\u001b[39mevaluate, op)\n\u001b[1;32m    219\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 220\u001b[0m     result \u001b[38;5;241m=\u001b[39m func(left, right)\n\u001b[1;32m    221\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m:\n\u001b[1;32m    222\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m is_cmp \u001b[38;5;129;01mand\u001b[39;00m (\n\u001b[1;32m    223\u001b[0m         left\u001b[38;5;241m.\u001b[39mdtype \u001b[38;5;241m==\u001b[39m \u001b[38;5;28mobject\u001b[39m \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mgetattr\u001b[39m(right, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdtype\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m) \u001b[38;5;241m==\u001b[39m \u001b[38;5;28mobject\u001b[39m\n\u001b[1;32m    224\u001b[0m     ):\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    227\u001b[0m         \u001b[38;5;66;03m# Don't do this for comparisons, as that will handle complex numbers\u001b[39;00m\n\u001b[1;32m    228\u001b[0m         \u001b[38;5;66;03m#  incorrectly, see GH#32047\u001b[39;00m\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/pandas/core/computation/expressions.py:242\u001b[0m, in \u001b[0;36mevaluate\u001b[0;34m(op, a, b, use_numexpr)\u001b[0m\n\u001b[1;32m    239\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m op_str \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m    240\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m use_numexpr:\n\u001b[1;32m    241\u001b[0m         \u001b[38;5;66;03m# error: \"None\" not callable\u001b[39;00m\n\u001b[0;32m--> 242\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m _evaluate(op, op_str, a, b)  \u001b[38;5;66;03m# type: ignore[misc]\u001b[39;00m\n\u001b[1;32m    243\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m _evaluate_standard(op, op_str, a, b)\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/pandas/core/computation/expressions.py:131\u001b[0m, in \u001b[0;36m_evaluate_numexpr\u001b[0;34m(op, op_str, a, b)\u001b[0m\n\u001b[1;32m    128\u001b[0m     _store_test_result(result \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m)\n\u001b[1;32m    130\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m result \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m--> 131\u001b[0m     result \u001b[38;5;241m=\u001b[39m _evaluate_standard(op, op_str, a, b)\n\u001b[1;32m    133\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m result\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/pandas/core/computation/expressions.py:73\u001b[0m, in \u001b[0;36m_evaluate_standard\u001b[0;34m(op, op_str, a, b)\u001b[0m\n\u001b[1;32m     71\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m _TEST_MODE:\n\u001b[1;32m     72\u001b[0m     _store_test_result(\u001b[38;5;28;01mFalse\u001b[39;00m)\n\u001b[0;32m---> 73\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m op(a, b)\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/pandas/core/roperator.py:19\u001b[0m, in \u001b[0;36mrmul\u001b[0;34m(left, right)\u001b[0m\n\u001b[1;32m     18\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mrmul\u001b[39m(left, right):\n\u001b[0;32m---> 19\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m right \u001b[38;5;241m*\u001b[39m left\n",
      "\u001b[0;31mValueError\u001b[0m: operands could not be broadcast together with shapes (237,) (259,) "
     ]
    }
   ],
   "source": [
    "N = 47\n",
    "M = 6\n",
    "idx = np.repeat(range(M-1), N)\n",
    "idx = np.append(idx, 5)\n",
    "idx = np.append(idx, 5)\n",
    "np.random.seed(42)\n",
    "\n",
    "alpha_real = np.random.normal(2.5, 0.5, size=M)\n",
    "beta_real = np.random.beta(6, 1, size=M)\n",
    "eps_real = np.random.normal(0, 0.5, size=len(idx))\n",
    "\n",
    "y_m = df.y\n",
    "x_m = df.x\n",
    "y_m = alpha_real[idx] + beta_real[idx] * x_m + eps_real\n",
    "\n",
    "_, ax = plt.subplots(2, 3, figsize=(10, 5), sharex=True, sharey=True)\n",
    "ax = np.ravel(ax)\n",
    "j, k = 0, N\n",
    "for i in range(M):\n",
    "    ax[i].scatter(x_m[j:k], y_m[j:k])\n",
    "    ax[i].set_xlabel(f'x_{i}')\n",
    "    ax[i].set_ylabel(f'y_{i}', rotation=0, labelpad=15)\n",
    "    j += N\n",
    "    k += N\n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 607
    },
    "id": "Ku-PejBpkQ6B",
    "outputId": "3acea58a-1a04-4688-8344-3627060c6d79"
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/rewriting/shape.py:157: UserWarning: Failed to infer_shape from Op Mul.\n",
      "Input shapes: [(TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),)]\n",
      "Exception encountered during infer_shape: <class 'ValueError'>\n",
      "Exception message: Could not broadcast dimensions. Incompatible shapes were [(TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),)].\n",
      "Traceback: Traceback (most recent call last):\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/rewriting/shape.py\", line 133, in get_node_infer_shape\n",
      "    o_shapes = shape_infer(\n",
      "               ^^^^^^^^^^^^\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/elemwise.py\", line 837, in infer_shape\n",
      "    out_shape = broadcast_shape(*i_shapes, arrays_are_shapes=True)\n",
      "                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/extra_ops.py\", line 1467, in broadcast_shape\n",
      "    return broadcast_shape_iter(arrays, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/extra_ops.py\", line 1548, in broadcast_shape_iter\n",
      "    raise ValueError(\n",
      "ValueError: Could not broadcast dimensions. Incompatible shapes were [(TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),)].\n",
      "\n",
      "  warn(msg)\n",
      "/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/rewriting/shape.py:157: UserWarning: Failed to infer_shape from Op Add.\n",
      "Input shapes: [(TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),)]\n",
      "Exception encountered during infer_shape: <class 'ValueError'>\n",
      "Exception message: Could not broadcast dimensions. Incompatible shapes were [(TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),)].\n",
      "Traceback: Traceback (most recent call last):\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/rewriting/shape.py\", line 133, in get_node_infer_shape\n",
      "    o_shapes = shape_infer(\n",
      "               ^^^^^^^^^^^^\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/elemwise.py\", line 837, in infer_shape\n",
      "    out_shape = broadcast_shape(*i_shapes, arrays_are_shapes=True)\n",
      "                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/extra_ops.py\", line 1467, in broadcast_shape\n",
      "    return broadcast_shape_iter(arrays, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/extra_ops.py\", line 1548, in broadcast_shape_iter\n",
      "    raise ValueError(\n",
      "ValueError: Could not broadcast dimensions. Incompatible shapes were [(TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),)].\n",
      "\n",
      "  warn(msg)\n",
      "/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/rewriting/shape.py:157: UserWarning: Failed to infer_shape from Op Mul.\n",
      "Input shapes: [(TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),)]\n",
      "Exception encountered during infer_shape: <class 'ValueError'>\n",
      "Exception message: Could not broadcast dimensions. Incompatible shapes were [(TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),)].\n",
      "Traceback: Traceback (most recent call last):\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/rewriting/shape.py\", line 133, in get_node_infer_shape\n",
      "    o_shapes = shape_infer(\n",
      "               ^^^^^^^^^^^^\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/elemwise.py\", line 837, in infer_shape\n",
      "    out_shape = broadcast_shape(*i_shapes, arrays_are_shapes=True)\n",
      "                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/extra_ops.py\", line 1467, in broadcast_shape\n",
      "    return broadcast_shape_iter(arrays, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/extra_ops.py\", line 1548, in broadcast_shape_iter\n",
      "    raise ValueError(\n",
      "ValueError: Could not broadcast dimensions. Incompatible shapes were [(TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),)].\n",
      "\n",
      "  warn(msg)\n",
      "/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/rewriting/shape.py:157: UserWarning: Failed to infer_shape from Op Add.\n",
      "Input shapes: [(TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),)]\n",
      "Exception encountered during infer_shape: <class 'ValueError'>\n",
      "Exception message: Could not broadcast dimensions. Incompatible shapes were [(TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),)].\n",
      "Traceback: Traceback (most recent call last):\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/rewriting/shape.py\", line 133, in get_node_infer_shape\n",
      "    o_shapes = shape_infer(\n",
      "               ^^^^^^^^^^^^\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/elemwise.py\", line 837, in infer_shape\n",
      "    out_shape = broadcast_shape(*i_shapes, arrays_are_shapes=True)\n",
      "                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/extra_ops.py\", line 1467, in broadcast_shape\n",
      "    return broadcast_shape_iter(arrays, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/extra_ops.py\", line 1548, in broadcast_shape_iter\n",
      "    raise ValueError(\n",
      "ValueError: Could not broadcast dimensions. Incompatible shapes were [(TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),)].\n",
      "\n",
      "  warn(msg)\n",
      "/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/rewriting/shape.py:157: UserWarning: Failed to infer_shape from Op Mul.\n",
      "Input shapes: [(TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),)]\n",
      "Exception encountered during infer_shape: <class 'ValueError'>\n",
      "Exception message: Could not broadcast dimensions. Incompatible shapes were [(TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),)].\n",
      "Traceback: Traceback (most recent call last):\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/rewriting/shape.py\", line 133, in get_node_infer_shape\n",
      "    o_shapes = shape_infer(\n",
      "               ^^^^^^^^^^^^\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/elemwise.py\", line 837, in infer_shape\n",
      "    out_shape = broadcast_shape(*i_shapes, arrays_are_shapes=True)\n",
      "                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/extra_ops.py\", line 1467, in broadcast_shape\n",
      "    return broadcast_shape_iter(arrays, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/extra_ops.py\", line 1548, in broadcast_shape_iter\n",
      "    raise ValueError(\n",
      "ValueError: Could not broadcast dimensions. Incompatible shapes were [(TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),)].\n",
      "\n",
      "  warn(msg)\n",
      "/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/rewriting/shape.py:157: UserWarning: Failed to infer_shape from Op Add.\n",
      "Input shapes: [(TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),)]\n",
      "Exception encountered during infer_shape: <class 'ValueError'>\n",
      "Exception message: Could not broadcast dimensions. Incompatible shapes were [(TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),)].\n",
      "Traceback: Traceback (most recent call last):\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/rewriting/shape.py\", line 133, in get_node_infer_shape\n",
      "    o_shapes = shape_infer(\n",
      "               ^^^^^^^^^^^^\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/elemwise.py\", line 837, in infer_shape\n",
      "    out_shape = broadcast_shape(*i_shapes, arrays_are_shapes=True)\n",
      "                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/extra_ops.py\", line 1467, in broadcast_shape\n",
      "    return broadcast_shape_iter(arrays, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/extra_ops.py\", line 1548, in broadcast_shape_iter\n",
      "    raise ValueError(\n",
      "ValueError: Could not broadcast dimensions. Incompatible shapes were [(TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),)].\n",
      "\n",
      "  warn(msg)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/rewriting/shape.py:157: UserWarning: Failed to infer_shape from Op Composite{...}.\n",
      "Input shapes: [(TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),), (TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),), (TensorConstant(TensorType(int64, shape=()), data=array(1)),), (TensorConstant(TensorType(int64, shape=()), data=array(1)),), (TensorConstant(TensorType(int64, shape=()), data=array(1)),), (TensorConstant(TensorType(int64, shape=()), data=array(1)),), (TensorConstant(TensorType(int64, shape=()), data=array(1)),), (TensorConstant(TensorType(int64, shape=()), data=array(1)),), (TensorConstant(TensorType(int64, shape=()), data=array(1)),), (TensorConstant(TensorType(int64, shape=()), data=array(1)),), (TensorConstant(TensorType(int64, shape=()), data=array(1)),), (TensorConstant(TensorType(int64, shape=()), data=array(1)),)]\n",
      "Exception encountered during infer_shape: <class 'ValueError'>\n",
      "Exception message: Could not broadcast dimensions. Incompatible shapes were [(TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),), (TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),), (ScalarConstant(ScalarType(int64), data=1),), (ScalarConstant(ScalarType(int64), data=1),), (ScalarConstant(ScalarType(int64), data=1),), (ScalarConstant(ScalarType(int64), data=1),), (ScalarConstant(ScalarType(int64), data=1),), (ScalarConstant(ScalarType(int64), data=1),), (ScalarConstant(ScalarType(int64), data=1),), (ScalarConstant(ScalarType(int64), data=1),), (ScalarConstant(ScalarType(int64), data=1),), (ScalarConstant(ScalarType(int64), data=1),)].\n",
      "Traceback: Traceback (most recent call last):\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/rewriting/shape.py\", line 133, in get_node_infer_shape\n",
      "    o_shapes = shape_infer(\n",
      "               ^^^^^^^^^^^^\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/elemwise.py\", line 837, in infer_shape\n",
      "    out_shape = broadcast_shape(*i_shapes, arrays_are_shapes=True)\n",
      "                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/extra_ops.py\", line 1467, in broadcast_shape\n",
      "    return broadcast_shape_iter(arrays, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/extra_ops.py\", line 1548, in broadcast_shape_iter\n",
      "    raise ValueError(\n",
      "ValueError: Could not broadcast dimensions. Incompatible shapes were [(TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),), (TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),), (ScalarConstant(ScalarType(int64), data=1),), (ScalarConstant(ScalarType(int64), data=1),), (ScalarConstant(ScalarType(int64), data=1),), (ScalarConstant(ScalarType(int64), data=1),), (ScalarConstant(ScalarType(int64), data=1),), (ScalarConstant(ScalarType(int64), data=1),), (ScalarConstant(ScalarType(int64), data=1),), (ScalarConstant(ScalarType(int64), data=1),), (ScalarConstant(ScalarType(int64), data=1),), (ScalarConstant(ScalarType(int64), data=1),)].\n",
      "\n",
      "  warn(msg)\n",
      "/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/rewriting/shape.py:157: UserWarning: Failed to infer_shape from Op Composite{...}.\n",
      "Input shapes: [(TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),), (TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),), (TensorConstant(TensorType(int64, shape=()), data=array(1)),), (TensorConstant(TensorType(int64, shape=()), data=array(1)),), (TensorConstant(TensorType(int64, shape=()), data=array(1)),), (TensorConstant(TensorType(int64, shape=()), data=array(1)),), (TensorConstant(TensorType(int64, shape=()), data=array(1)),), (TensorConstant(TensorType(int64, shape=()), data=array(1)),), (TensorConstant(TensorType(int64, shape=()), data=array(1)),)]\n",
      "Exception encountered during infer_shape: <class 'ValueError'>\n",
      "Exception message: Could not broadcast dimensions. Incompatible shapes were [(TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),), (TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),), (ScalarConstant(ScalarType(int64), data=1),), (ScalarConstant(ScalarType(int64), data=1),), (ScalarConstant(ScalarType(int64), data=1),), (ScalarConstant(ScalarType(int64), data=1),), (ScalarConstant(ScalarType(int64), data=1),), (ScalarConstant(ScalarType(int64), data=1),), (ScalarConstant(ScalarType(int64), data=1),)].\n",
      "Traceback: Traceback (most recent call last):\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/rewriting/shape.py\", line 133, in get_node_infer_shape\n",
      "    o_shapes = shape_infer(\n",
      "               ^^^^^^^^^^^^\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/elemwise.py\", line 837, in infer_shape\n",
      "    out_shape = broadcast_shape(*i_shapes, arrays_are_shapes=True)\n",
      "                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/extra_ops.py\", line 1467, in broadcast_shape\n",
      "    return broadcast_shape_iter(arrays, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/extra_ops.py\", line 1548, in broadcast_shape_iter\n",
      "    raise ValueError(\n",
      "ValueError: Could not broadcast dimensions. Incompatible shapes were [(TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),), (TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),), (ScalarConstant(ScalarType(int64), data=1),), (ScalarConstant(ScalarType(int64), data=1),), (ScalarConstant(ScalarType(int64), data=1),), (ScalarConstant(ScalarType(int64), data=1),), (ScalarConstant(ScalarType(int64), data=1),), (ScalarConstant(ScalarType(int64), data=1),), (ScalarConstant(ScalarType(int64), data=1),)].\n",
      "\n",
      "  warn(msg)\n",
      "Auto-assigning NUTS sampler...\n",
      "Initializing NUTS using jitter+adapt_diag...\n",
      "/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/rewriting/shape.py:157: UserWarning: Failed to infer_shape from Op Mul.\n",
      "Input shapes: [(TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),)]\n",
      "Exception encountered during infer_shape: <class 'ValueError'>\n",
      "Exception message: Could not broadcast dimensions. Incompatible shapes were [(TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),)].\n",
      "Traceback: Traceback (most recent call last):\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/rewriting/shape.py\", line 133, in get_node_infer_shape\n",
      "    o_shapes = shape_infer(\n",
      "               ^^^^^^^^^^^^\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/elemwise.py\", line 837, in infer_shape\n",
      "    out_shape = broadcast_shape(*i_shapes, arrays_are_shapes=True)\n",
      "                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/extra_ops.py\", line 1467, in broadcast_shape\n",
      "    return broadcast_shape_iter(arrays, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/extra_ops.py\", line 1548, in broadcast_shape_iter\n",
      "    raise ValueError(\n",
      "ValueError: Could not broadcast dimensions. Incompatible shapes were [(TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),)].\n",
      "\n",
      "  warn(msg)\n",
      "/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/rewriting/shape.py:157: UserWarning: Failed to infer_shape from Op Add.\n",
      "Input shapes: [(TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),)]\n",
      "Exception encountered during infer_shape: <class 'ValueError'>\n",
      "Exception message: Could not broadcast dimensions. Incompatible shapes were [(TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),)].\n",
      "Traceback: Traceback (most recent call last):\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/rewriting/shape.py\", line 133, in get_node_infer_shape\n",
      "    o_shapes = shape_infer(\n",
      "               ^^^^^^^^^^^^\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/elemwise.py\", line 837, in infer_shape\n",
      "    out_shape = broadcast_shape(*i_shapes, arrays_are_shapes=True)\n",
      "                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/extra_ops.py\", line 1467, in broadcast_shape\n",
      "    return broadcast_shape_iter(arrays, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/extra_ops.py\", line 1548, in broadcast_shape_iter\n",
      "    raise ValueError(\n",
      "ValueError: Could not broadcast dimensions. Incompatible shapes were [(TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),)].\n",
      "\n",
      "  warn(msg)\n",
      "/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/rewriting/shape.py:157: UserWarning: Failed to infer_shape from Op Mul.\n",
      "Input shapes: [(TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),)]\n",
      "Exception encountered during infer_shape: <class 'ValueError'>\n",
      "Exception message: Could not broadcast dimensions. Incompatible shapes were [(TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),)].\n",
      "Traceback: Traceback (most recent call last):\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/rewriting/shape.py\", line 133, in get_node_infer_shape\n",
      "    o_shapes = shape_infer(\n",
      "               ^^^^^^^^^^^^\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/elemwise.py\", line 837, in infer_shape\n",
      "    out_shape = broadcast_shape(*i_shapes, arrays_are_shapes=True)\n",
      "                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/extra_ops.py\", line 1467, in broadcast_shape\n",
      "    return broadcast_shape_iter(arrays, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/extra_ops.py\", line 1548, in broadcast_shape_iter\n",
      "    raise ValueError(\n",
      "ValueError: Could not broadcast dimensions. Incompatible shapes were [(TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),)].\n",
      "\n",
      "  warn(msg)\n",
      "/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/rewriting/shape.py:157: UserWarning: Failed to infer_shape from Op Add.\n",
      "Input shapes: [(TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),)]\n",
      "Exception encountered during infer_shape: <class 'ValueError'>\n",
      "Exception message: Could not broadcast dimensions. Incompatible shapes were [(TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),)].\n",
      "Traceback: Traceback (most recent call last):\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/rewriting/shape.py\", line 133, in get_node_infer_shape\n",
      "    o_shapes = shape_infer(\n",
      "               ^^^^^^^^^^^^\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/elemwise.py\", line 837, in infer_shape\n",
      "    out_shape = broadcast_shape(*i_shapes, arrays_are_shapes=True)\n",
      "                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/extra_ops.py\", line 1467, in broadcast_shape\n",
      "    return broadcast_shape_iter(arrays, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/extra_ops.py\", line 1548, in broadcast_shape_iter\n",
      "    raise ValueError(\n",
      "ValueError: Could not broadcast dimensions. Incompatible shapes were [(TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),)].\n",
      "\n",
      "  warn(msg)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/rewriting/shape.py:157: UserWarning: Failed to infer_shape from Op Composite{switch(i9, (i8 - (i6 * i7 * log1p(((i4 * sqr((i3 - (i2 + (i0 * i1))))) / i5)))), i10)}.\n",
      "Input shapes: [(TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),), (TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),), (TensorConstant(TensorType(int64, shape=()), data=array(1)),), (TensorConstant(TensorType(int64, shape=()), data=array(1)),), (TensorConstant(TensorType(int64, shape=()), data=array(1)),), (TensorConstant(TensorType(int64, shape=()), data=array(1)),), (TensorConstant(TensorType(int64, shape=()), data=array(1)),), (TensorConstant(TensorType(int64, shape=()), data=array(1)),), (TensorConstant(TensorType(int64, shape=()), data=array(1)),)]\n",
      "Exception encountered during infer_shape: <class 'ValueError'>\n",
      "Exception message: Could not broadcast dimensions. Incompatible shapes were [(TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),), (TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),), (ScalarConstant(ScalarType(int64), data=1),), (ScalarConstant(ScalarType(int64), data=1),), (ScalarConstant(ScalarType(int64), data=1),), (ScalarConstant(ScalarType(int64), data=1),), (ScalarConstant(ScalarType(int64), data=1),), (ScalarConstant(ScalarType(int64), data=1),), (ScalarConstant(ScalarType(int64), data=1),)].\n",
      "Traceback: Traceback (most recent call last):\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/rewriting/shape.py\", line 133, in get_node_infer_shape\n",
      "    o_shapes = shape_infer(\n",
      "               ^^^^^^^^^^^^\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/elemwise.py\", line 837, in infer_shape\n",
      "    out_shape = broadcast_shape(*i_shapes, arrays_are_shapes=True)\n",
      "                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/extra_ops.py\", line 1467, in broadcast_shape\n",
      "    return broadcast_shape_iter(arrays, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/extra_ops.py\", line 1548, in broadcast_shape_iter\n",
      "    raise ValueError(\n",
      "ValueError: Could not broadcast dimensions. Incompatible shapes were [(TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),), (TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),), (ScalarConstant(ScalarType(int64), data=1),), (ScalarConstant(ScalarType(int64), data=1),), (ScalarConstant(ScalarType(int64), data=1),), (ScalarConstant(ScalarType(int64), data=1),), (ScalarConstant(ScalarType(int64), data=1),), (ScalarConstant(ScalarType(int64), data=1),), (ScalarConstant(ScalarType(int64), data=1),)].\n",
      "\n",
      "  warn(msg)\n",
      "/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/rewriting/shape.py:157: UserWarning: Failed to infer_shape from Op Composite{switch(i8, (i7 - (0.5 * i6 * log1p(((i4 * sqr((i3 - (i2 + (i0 * i1))))) / i5)))), -inf)}.\n",
      "Input shapes: [(TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),), (TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),), (TensorConstant(TensorType(int64, shape=()), data=array(1)),), (TensorConstant(TensorType(int64, shape=()), data=array(1)),), (TensorConstant(TensorType(int64, shape=()), data=array(1)),), (TensorConstant(TensorType(int64, shape=()), data=array(1)),), (TensorConstant(TensorType(int64, shape=()), data=array(1)),)]\n",
      "Exception encountered during infer_shape: <class 'ValueError'>\n",
      "Exception message: Could not broadcast dimensions. Incompatible shapes were [(TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),), (TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),), (ScalarConstant(ScalarType(int64), data=1),), (ScalarConstant(ScalarType(int64), data=1),), (ScalarConstant(ScalarType(int64), data=1),), (ScalarConstant(ScalarType(int64), data=1),), (ScalarConstant(ScalarType(int64), data=1),)].\n",
      "Traceback: Traceback (most recent call last):\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/rewriting/shape.py\", line 133, in get_node_infer_shape\n",
      "    o_shapes = shape_infer(\n",
      "               ^^^^^^^^^^^^\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/elemwise.py\", line 837, in infer_shape\n",
      "    out_shape = broadcast_shape(*i_shapes, arrays_are_shapes=True)\n",
      "                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/extra_ops.py\", line 1467, in broadcast_shape\n",
      "    return broadcast_shape_iter(arrays, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/imachome/anaconda3/lib/python3.11/site-packages/pytensor/tensor/extra_ops.py\", line 1548, in broadcast_shape_iter\n",
      "    raise ValueError(\n",
      "ValueError: Could not broadcast dimensions. Incompatible shapes were [(TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),), (TensorConstant(TensorType(int64, shape=()), data=array(237)),), (TensorConstant(TensorType(int64, shape=()), data=array(259)),), (ScalarConstant(ScalarType(int64), data=1),), (ScalarConstant(ScalarType(int64), data=1),), (ScalarConstant(ScalarType(int64), data=1),), (ScalarConstant(ScalarType(int64), data=1),), (ScalarConstant(ScalarType(int64), data=1),)].\n",
      "\n",
      "  warn(msg)\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "Input dimension mismatch: (input[0].shape[0] = 237, input[1].shape[0] = 259)\nApply node that caused the error: Composite{switch(i8, (i7 - (0.5 * i6 * log1p(((i4 * sqr((i3 - (i2 + (i0 * i1))))) / i5)))), -inf)}(AdvancedSubtensor1.0, [-1.702839 ... 85513e+00], AdvancedSubtensor1.0, y_pred{[ 9.064981 ... .43994714]}, ExpandDims{axis=0}.0, ExpandDims{axis=0}.0, Composite{...}.1, Composite{...}.0, ExpandDims{axis=0}.0)\nToposort index: 16\nInputs types: [TensorType(float64, shape=(None,)), TensorType(float64, shape=(259,)), TensorType(float64, shape=(None,)), TensorType(float64, shape=(259,)), TensorType(float64, shape=(1,)), TensorType(float64, shape=(1,)), TensorType(float64, shape=(1,)), TensorType(float64, shape=(1,)), TensorType(bool, shape=(1,))]\nInputs shapes: [(237,), (259,), (237,), (259,), (1,), (1,), (1,), (1,), (1,)]\nInputs strides: [(8,), (8,), (8,), (8,), (8,), (8,), (8,), (8,), (1,)]\nInputs values: ['not shown', 'not shown', 'not shown', 'not shown', array([0.00768707]), array([44.18783593]), array([45.18783593]), array([-3.3587036]), array([ True])]\nOutputs clients: [[Sum{axes=None}(lam > 0, nu > 0)]]\n\nHINT: Re-running with most PyTensor optimizations disabled could provide a back-trace showing when this node was created. This can be done by setting the PyTensor flag 'optimizer=fast_compile'. If that does not work, PyTensor optimizations can be disabled with 'optimizer=None'.\nHINT: Use the PyTensor flag `exception_verbosity=high` for a debug print-out and storage map footprint of this Apply node.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/pytensor/compile/function/types.py:970\u001b[0m, in \u001b[0;36mFunction.__call__\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    968\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    969\u001b[0m     outputs \u001b[38;5;241m=\u001b[39m (\n\u001b[0;32m--> 970\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mvm()\n\u001b[1;32m    971\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m output_subset \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m    972\u001b[0m         \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mvm(output_subset\u001b[38;5;241m=\u001b[39moutput_subset)\n\u001b[1;32m    973\u001b[0m     )\n\u001b[1;32m    974\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m:\n",
      "\u001b[0;31mValueError\u001b[0m: Input dimension mismatch: (input[0].shape[0] = 237, input[1].shape[0] = 259)",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[12], line 14\u001b[0m\n\u001b[1;32m      9\u001b[0m     y_pred \u001b[38;5;241m=\u001b[39m pm\u001b[38;5;241m.\u001b[39mStudentT(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124my_pred\u001b[39m\u001b[38;5;124m'\u001b[39m, mu\u001b[38;5;241m=\u001b[39mα_tmp[idx] \u001b[38;5;241m+\u001b[39m β[idx] \u001b[38;5;241m*\u001b[39m x_centered,\n\u001b[1;32m     10\u001b[0m                          sigma\u001b[38;5;241m=\u001b[39mε, nu\u001b[38;5;241m=\u001b[39mν, observed\u001b[38;5;241m=\u001b[39my_m)\n\u001b[1;32m     12\u001b[0m     α \u001b[38;5;241m=\u001b[39m pm\u001b[38;5;241m.\u001b[39mDeterministic(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mα\u001b[39m\u001b[38;5;124m'\u001b[39m, α_tmp \u001b[38;5;241m-\u001b[39m β \u001b[38;5;241m*\u001b[39m x_m\u001b[38;5;241m.\u001b[39mmean())\n\u001b[0;32m---> 14\u001b[0m     trace_up \u001b[38;5;241m=\u001b[39m pm\u001b[38;5;241m.\u001b[39msample(\u001b[38;5;241m2000\u001b[39m)\n\u001b[1;32m     15\u001b[0m az\u001b[38;5;241m.\u001b[39mplot_forest(trace_up, var_names\u001b[38;5;241m=\u001b[39m[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mα\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mβ\u001b[39m\u001b[38;5;124m'\u001b[39m], combined\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/pymc/sampling/mcmc.py:682\u001b[0m, in \u001b[0;36msample\u001b[0;34m(draws, tune, chains, cores, random_seed, progressbar, step, nuts_sampler, initvals, init, jitter_max_retries, n_init, trace, discard_tuned_samples, compute_convergence_checks, keep_warning_stat, return_inferencedata, idata_kwargs, nuts_sampler_kwargs, callback, mp_ctx, model, **kwargs)\u001b[0m\n\u001b[1;32m    680\u001b[0m         [kwargs\u001b[38;5;241m.\u001b[39msetdefault(k, v) \u001b[38;5;28;01mfor\u001b[39;00m k, v \u001b[38;5;129;01min\u001b[39;00m nuts_kwargs\u001b[38;5;241m.\u001b[39mitems()]\n\u001b[1;32m    681\u001b[0m     _log\u001b[38;5;241m.\u001b[39minfo(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mAuto-assigning NUTS sampler...\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m--> 682\u001b[0m     initial_points, step \u001b[38;5;241m=\u001b[39m init_nuts(\n\u001b[1;32m    683\u001b[0m         init\u001b[38;5;241m=\u001b[39minit,\n\u001b[1;32m    684\u001b[0m         chains\u001b[38;5;241m=\u001b[39mchains,\n\u001b[1;32m    685\u001b[0m         n_init\u001b[38;5;241m=\u001b[39mn_init,\n\u001b[1;32m    686\u001b[0m         model\u001b[38;5;241m=\u001b[39mmodel,\n\u001b[1;32m    687\u001b[0m         random_seed\u001b[38;5;241m=\u001b[39mrandom_seed_list,\n\u001b[1;32m    688\u001b[0m         progressbar\u001b[38;5;241m=\u001b[39mprogressbar,\n\u001b[1;32m    689\u001b[0m         jitter_max_retries\u001b[38;5;241m=\u001b[39mjitter_max_retries,\n\u001b[1;32m    690\u001b[0m         tune\u001b[38;5;241m=\u001b[39mtune,\n\u001b[1;32m    691\u001b[0m         initvals\u001b[38;5;241m=\u001b[39minitvals,\n\u001b[1;32m    692\u001b[0m         \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs,\n\u001b[1;32m    693\u001b[0m     )\n\u001b[1;32m    695\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m initial_points \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m    696\u001b[0m     \u001b[38;5;66;03m# Time to draw/evaluate numeric start points for each chain.\u001b[39;00m\n\u001b[1;32m    697\u001b[0m     ipfns \u001b[38;5;241m=\u001b[39m make_initial_point_fns_per_chain(\n\u001b[1;32m    698\u001b[0m         model\u001b[38;5;241m=\u001b[39mmodel,\n\u001b[1;32m    699\u001b[0m         overrides\u001b[38;5;241m=\u001b[39minitvals,\n\u001b[1;32m    700\u001b[0m         jitter_rvs\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mset\u001b[39m(),\n\u001b[1;32m    701\u001b[0m         chains\u001b[38;5;241m=\u001b[39mchains,\n\u001b[1;32m    702\u001b[0m     )\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/pymc/sampling/mcmc.py:1327\u001b[0m, in \u001b[0;36minit_nuts\u001b[0;34m(init, chains, n_init, model, random_seed, progressbar, jitter_max_retries, tune, initvals, **kwargs)\u001b[0m\n\u001b[1;32m   1320\u001b[0m _log\u001b[38;5;241m.\u001b[39minfo(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mInitializing NUTS using \u001b[39m\u001b[38;5;132;01m{\u001b[39;00minit\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m...\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m   1322\u001b[0m cb \u001b[38;5;241m=\u001b[39m [\n\u001b[1;32m   1323\u001b[0m     pm\u001b[38;5;241m.\u001b[39mcallbacks\u001b[38;5;241m.\u001b[39mCheckParametersConvergence(tolerance\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m1e-2\u001b[39m, diff\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mabsolute\u001b[39m\u001b[38;5;124m\"\u001b[39m),\n\u001b[1;32m   1324\u001b[0m     pm\u001b[38;5;241m.\u001b[39mcallbacks\u001b[38;5;241m.\u001b[39mCheckParametersConvergence(tolerance\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m1e-2\u001b[39m, diff\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mrelative\u001b[39m\u001b[38;5;124m\"\u001b[39m),\n\u001b[1;32m   1325\u001b[0m ]\n\u001b[0;32m-> 1327\u001b[0m initial_points \u001b[38;5;241m=\u001b[39m _init_jitter(\n\u001b[1;32m   1328\u001b[0m     model,\n\u001b[1;32m   1329\u001b[0m     initvals,\n\u001b[1;32m   1330\u001b[0m     seeds\u001b[38;5;241m=\u001b[39mrandom_seed_list,\n\u001b[1;32m   1331\u001b[0m     jitter\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mjitter\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01min\u001b[39;00m init,\n\u001b[1;32m   1332\u001b[0m     jitter_max_retries\u001b[38;5;241m=\u001b[39mjitter_max_retries,\n\u001b[1;32m   1333\u001b[0m )\n\u001b[1;32m   1335\u001b[0m apoints \u001b[38;5;241m=\u001b[39m [DictToArrayBijection\u001b[38;5;241m.\u001b[39mmap(point) \u001b[38;5;28;01mfor\u001b[39;00m point \u001b[38;5;129;01min\u001b[39;00m initial_points]\n\u001b[1;32m   1336\u001b[0m apoints_data \u001b[38;5;241m=\u001b[39m [apoint\u001b[38;5;241m.\u001b[39mdata \u001b[38;5;28;01mfor\u001b[39;00m apoint \u001b[38;5;129;01min\u001b[39;00m apoints]\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/pymc/sampling/mcmc.py:1221\u001b[0m, in \u001b[0;36m_init_jitter\u001b[0;34m(model, initvals, seeds, jitter, jitter_max_retries)\u001b[0m\n\u001b[1;32m   1219\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m i \u001b[38;5;241m<\u001b[39m jitter_max_retries:\n\u001b[1;32m   1220\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 1221\u001b[0m         model\u001b[38;5;241m.\u001b[39mcheck_start_vals(point)\n\u001b[1;32m   1222\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m SamplingError:\n\u001b[1;32m   1223\u001b[0m         \u001b[38;5;66;03m# Retry with a new seed\u001b[39;00m\n\u001b[1;32m   1224\u001b[0m         seed \u001b[38;5;241m=\u001b[39m rng\u001b[38;5;241m.\u001b[39mrandint(\u001b[38;5;241m2\u001b[39m\u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39m\u001b[38;5;241m30\u001b[39m, dtype\u001b[38;5;241m=\u001b[39mnp\u001b[38;5;241m.\u001b[39mint64)\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/pymc/model.py:1697\u001b[0m, in \u001b[0;36mModel.check_start_vals\u001b[0;34m(self, start)\u001b[0m\n\u001b[1;32m   1691\u001b[0m     valid_keys \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m, \u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;241m.\u001b[39mjoin(value_names_set)\n\u001b[1;32m   1692\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(\n\u001b[1;32m   1693\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mSome start parameters do not appear in the model!\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   1694\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mValid keys are: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mvalid_keys\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m, but \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mextra_keys\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m was supplied\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   1695\u001b[0m     )\n\u001b[0;32m-> 1697\u001b[0m initial_eval \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mpoint_logps(point\u001b[38;5;241m=\u001b[39melem)\n\u001b[1;32m   1699\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28mall\u001b[39m(np\u001b[38;5;241m.\u001b[39misfinite(v) \u001b[38;5;28;01mfor\u001b[39;00m v \u001b[38;5;129;01min\u001b[39;00m initial_eval\u001b[38;5;241m.\u001b[39mvalues()):\n\u001b[1;32m   1700\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m SamplingError(\n\u001b[1;32m   1701\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mInitial evaluation of model at starting point failed!\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   1702\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mStarting values:\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;132;01m{\u001b[39;00melem\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   1703\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mLogp initial evaluation results:\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;132;01m{\u001b[39;00minitial_eval\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   1704\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mYou can call `model.debug()` for more details.\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   1705\u001b[0m     )\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/pymc/model.py:1732\u001b[0m, in \u001b[0;36mModel.point_logps\u001b[0;34m(self, point, round_vals)\u001b[0m\n\u001b[1;32m   1726\u001b[0m factors \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mbasic_RVs \u001b[38;5;241m+\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mpotentials\n\u001b[1;32m   1727\u001b[0m factor_logps_fn \u001b[38;5;241m=\u001b[39m [pt\u001b[38;5;241m.\u001b[39msum(factor) \u001b[38;5;28;01mfor\u001b[39;00m factor \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mlogp(factors, \u001b[38;5;28msum\u001b[39m\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m)]\n\u001b[1;32m   1728\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m {\n\u001b[1;32m   1729\u001b[0m     factor\u001b[38;5;241m.\u001b[39mname: np\u001b[38;5;241m.\u001b[39mround(np\u001b[38;5;241m.\u001b[39masarray(factor_logp), round_vals)\n\u001b[1;32m   1730\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m factor, factor_logp \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mzip\u001b[39m(\n\u001b[1;32m   1731\u001b[0m         factors,\n\u001b[0;32m-> 1732\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcompile_fn(factor_logps_fn)(point),\n\u001b[1;32m   1733\u001b[0m     )\n\u001b[1;32m   1734\u001b[0m }\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/pymc/pytensorf.py:762\u001b[0m, in \u001b[0;36mPointFunc.__call__\u001b[0;34m(self, state)\u001b[0m\n\u001b[1;32m    761\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m__call__\u001b[39m(\u001b[38;5;28mself\u001b[39m, state):\n\u001b[0;32m--> 762\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mf(\u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mstate)\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/pytensor/compile/function/types.py:983\u001b[0m, in \u001b[0;36mFunction.__call__\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    981\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mhasattr\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mvm, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mthunks\u001b[39m\u001b[38;5;124m\"\u001b[39m):\n\u001b[1;32m    982\u001b[0m         thunk \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mvm\u001b[38;5;241m.\u001b[39mthunks[\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mvm\u001b[38;5;241m.\u001b[39mposition_of_error]\n\u001b[0;32m--> 983\u001b[0m     raise_with_op(\n\u001b[1;32m    984\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mmaker\u001b[38;5;241m.\u001b[39mfgraph,\n\u001b[1;32m    985\u001b[0m         node\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mvm\u001b[38;5;241m.\u001b[39mnodes[\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mvm\u001b[38;5;241m.\u001b[39mposition_of_error],\n\u001b[1;32m    986\u001b[0m         thunk\u001b[38;5;241m=\u001b[39mthunk,\n\u001b[1;32m    987\u001b[0m         storage_map\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mgetattr\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mvm, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mstorage_map\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m),\n\u001b[1;32m    988\u001b[0m     )\n\u001b[1;32m    989\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    990\u001b[0m     \u001b[38;5;66;03m# old-style linkers raise their own exceptions\u001b[39;00m\n\u001b[1;32m    991\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/pytensor/link/utils.py:535\u001b[0m, in \u001b[0;36mraise_with_op\u001b[0;34m(fgraph, node, thunk, exc_info, storage_map)\u001b[0m\n\u001b[1;32m    530\u001b[0m     warnings\u001b[38;5;241m.\u001b[39mwarn(\n\u001b[1;32m    531\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mexc_type\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m error does not allow us to add an extra error message\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    532\u001b[0m     )\n\u001b[1;32m    533\u001b[0m     \u001b[38;5;66;03m# Some exception need extra parameter in inputs. So forget the\u001b[39;00m\n\u001b[1;32m    534\u001b[0m     \u001b[38;5;66;03m# extra long error message in that case.\u001b[39;00m\n\u001b[0;32m--> 535\u001b[0m \u001b[38;5;28;01mraise\u001b[39;00m exc_value\u001b[38;5;241m.\u001b[39mwith_traceback(exc_trace)\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/pytensor/compile/function/types.py:970\u001b[0m, in \u001b[0;36mFunction.__call__\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    967\u001b[0m t0_fn \u001b[38;5;241m=\u001b[39m time\u001b[38;5;241m.\u001b[39mperf_counter()\n\u001b[1;32m    968\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    969\u001b[0m     outputs \u001b[38;5;241m=\u001b[39m (\n\u001b[0;32m--> 970\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mvm()\n\u001b[1;32m    971\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m output_subset \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m    972\u001b[0m         \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mvm(output_subset\u001b[38;5;241m=\u001b[39moutput_subset)\n\u001b[1;32m    973\u001b[0m     )\n\u001b[1;32m    974\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m:\n\u001b[1;32m    975\u001b[0m     restore_defaults()\n",
      "\u001b[0;31mValueError\u001b[0m: Input dimension mismatch: (input[0].shape[0] = 237, input[1].shape[0] = 259)\nApply node that caused the error: Composite{switch(i8, (i7 - (0.5 * i6 * log1p(((i4 * sqr((i3 - (i2 + (i0 * i1))))) / i5)))), -inf)}(AdvancedSubtensor1.0, [-1.702839 ... 85513e+00], AdvancedSubtensor1.0, y_pred{[ 9.064981 ... .43994714]}, ExpandDims{axis=0}.0, ExpandDims{axis=0}.0, Composite{...}.1, Composite{...}.0, ExpandDims{axis=0}.0)\nToposort index: 16\nInputs types: [TensorType(float64, shape=(None,)), TensorType(float64, shape=(259,)), TensorType(float64, shape=(None,)), TensorType(float64, shape=(259,)), TensorType(float64, shape=(1,)), TensorType(float64, shape=(1,)), TensorType(float64, shape=(1,)), TensorType(float64, shape=(1,)), TensorType(bool, shape=(1,))]\nInputs shapes: [(237,), (259,), (237,), (259,), (1,), (1,), (1,), (1,), (1,)]\nInputs strides: [(8,), (8,), (8,), (8,), (8,), (8,), (8,), (8,), (1,)]\nInputs values: ['not shown', 'not shown', 'not shown', 'not shown', array([0.00768707]), array([44.18783593]), array([45.18783593]), array([-3.3587036]), array([ True])]\nOutputs clients: [[Sum{axes=None}(lam > 0, nu > 0)]]\n\nHINT: Re-running with most PyTensor optimizations disabled could provide a back-trace showing when this node was created. This can be done by setting the PyTensor flag 'optimizer=fast_compile'. If that does not work, PyTensor optimizations can be disabled with 'optimizer=None'.\nHINT: Use the PyTensor flag `exception_verbosity=high` for a debug print-out and storage map footprint of this Apply node."
     ]
    }
   ],
   "source": [
    "x_centered = x_m - x_m.mean()\n",
    "\n",
    "with pm.Model() as unpooled_model:\n",
    "    α_tmp = pm.Normal('α_tmp', mu=0, sigma=10, shape=M)\n",
    "    β = pm.Normal('β', mu=0, sigma=10, shape=M)\n",
    "    ε = pm.HalfCauchy('ε', 5)\n",
    "    ν = pm.Exponential('ν', 1/30)\n",
    "\n",
    "    y_pred = pm.StudentT('y_pred', mu=α_tmp[idx] + β[idx] * x_centered,\n",
    "                         sigma=ε, nu=ν, observed=y_m)\n",
    "\n",
    "    α = pm.Deterministic('α', α_tmp - β * x_m.mean())\n",
    "\n",
    "    trace_up = pm.sample(2000)\n",
    "az.plot_forest(trace_up, var_names=['α', 'β'], combined=True);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 607
    },
    "id": "sUg0lV-jkRMC",
    "outputId": "28024fe7-b560-4e61-a8f2-8bc56eca1dbc"
   },
   "outputs": [],
   "source": [
    "with pm.Model() as hierarchical_model:\n",
    "    # априорные гипер-распределения\n",
    "    α_μ_tmp = pm.Normal('α_μ_tmp', mu=0, sigma=10)\n",
    "    α_σ_tmp = pm.HalfNormal('α_σ_tmp', sigma=10)\n",
    "    β_μ = pm.Normal('β_μ', mu=0, sigma=1)\n",
    "    β_σ = pm.HalfNormal('β_σ', sigma=1)\n",
    "\n",
    "    # априорные распеределения\n",
    "    α_tmp = pm.Normal('α_tmp', mu=α_μ_tmp, sigma=α_σ_tmp, shape=M)\n",
    "    β = pm.Normal('β', mu=β_μ, sigma=β_σ, shape=M)\n",
    "    ε = pm.HalfCauchy('ε', 5)\n",
    "    ν = pm.Exponential('ν', 1/30)\n",
    "\n",
    "    # апосториорное распределение\n",
    "    y_pred = pm.StudentT('y_pred', mu=α_tmp[idx] + β[idx] * x_centered,\n",
    "                         sigma=ε, nu=ν, observed=y_m)\n",
    "\n",
    "    # обратное перемасштабирование\n",
    "    α = pm.Deterministic('α', α_tmp - β * x_m.mean())\n",
    "    α_μ = pm.Deterministic('α_μ', α_μ_tmp - β_μ * x_m.mean())\n",
    "    α_σ = pm.Deterministic('α_sd', α_σ_tmp - β_μ * x_m.mean())\n",
    "\n",
    "    # трассировка\n",
    "    trace_hm = pm.sample(3000,tune = 1000) #target_accept=.8\n",
    "\n",
    "az.plot_forest(trace_hm, var_names=['α', 'β'], combined=True);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 633
    },
    "id": "Zn_JvNsqCr9o",
    "outputId": "c41aef2d-3b08-473d-b9c7-e0005ce87554"
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'trace_hm' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[10], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m az\u001b[38;5;241m.\u001b[39mplot_forest(trace_hm, var_names\u001b[38;5;241m=\u001b[39m[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mα\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mβ\u001b[39m\u001b[38;5;124m'\u001b[39m],kind \u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mridgeplot\u001b[39m\u001b[38;5;124m'\u001b[39m,\n\u001b[1;32m      2\u001b[0m                combined\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m,\n\u001b[1;32m      3\u001b[0m                ridgeplot_truncate\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m,\n\u001b[1;32m      4\u001b[0m                ridgeplot_quantiles\u001b[38;5;241m=\u001b[39m[\u001b[38;5;241m.25\u001b[39m, \u001b[38;5;241m.5\u001b[39m, \u001b[38;5;241m.75\u001b[39m],\n\u001b[1;32m      5\u001b[0m                ridgeplot_overlap\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0.7\u001b[39m,\n\u001b[1;32m      6\u001b[0m                colors\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mwhite\u001b[39m\u001b[38;5;124m'\u001b[39m)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'trace_hm' is not defined"
     ]
    }
   ],
   "source": [
    "az.plot_forest(trace_hm, var_names=['α', 'β'],kind ='ridgeplot',\n",
    "               combined=True,\n",
    "               ridgeplot_truncate=False,\n",
    "               ridgeplot_quantiles=[.25, .5, .75],\n",
    "               ridgeplot_overlap=0.7,\n",
    "               colors='white');"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 928
    },
    "id": "eLF-ZJR_BauI",
    "outputId": "4edebea7-d3e2-4716-97e0-5bd83c6317da"
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'trace_hm' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[11], line 4\u001b[0m\n\u001b[1;32m      1\u001b[0m _, ax \u001b[38;5;241m=\u001b[39m plt\u001b[38;5;241m.\u001b[39msubplots(\u001b[38;5;241m2\u001b[39m, \u001b[38;5;241m3\u001b[39m, figsize\u001b[38;5;241m=\u001b[39m(\u001b[38;5;241m15\u001b[39m, \u001b[38;5;241m10\u001b[39m), sharex\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m, sharey\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m,\n\u001b[1;32m      2\u001b[0m                      constrained_layout\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[0;32m----> 4\u001b[0m trace_hmm \u001b[38;5;241m=\u001b[39m trace_hm\u001b[38;5;241m.\u001b[39mposterior\u001b[38;5;241m.\u001b[39mstack(samples\u001b[38;5;241m=\u001b[39m{\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mchain\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdraw\u001b[39m\u001b[38;5;124m\"\u001b[39m})\n\u001b[1;32m      5\u001b[0m trace_upp \u001b[38;5;241m=\u001b[39m trace_up\u001b[38;5;241m.\u001b[39mposterior\u001b[38;5;241m.\u001b[39mstack(samples\u001b[38;5;241m=\u001b[39m{\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mchain\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdraw\u001b[39m\u001b[38;5;124m\"\u001b[39m})\n\u001b[1;32m      8\u001b[0m ax \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mravel(ax)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'trace_hm' is not defined"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABecAAAPzCAYAAAA9DLdyAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy81sbWrAAAACXBIWXMAAA9hAAAPYQGoP6dpAABIKklEQVR4nO3df2zX9bX48QOtjS2BkaUGsl0ylgoyRpVKWbdlBK91caKAu2O6u93k7u5mc+m1Wp1kc+QuiitzP92ts3fkLgszY85cog4nYtziHNnlUp0Ye+XKAC/Iwr1eYUMuFFb74fP9Y7fc2y+4y7utp2v7eCQm65v3x8+LA2+PPO3aCeVyuRwAAAAAAECaiSN9AAAAAAAAGG/EeQAAAAAASCbOAwAAAABAMnEeAAAAAACSifMAAAAAAJBMnAcAAAAAgGTiPAAAAAAAJBPnAQAAAAAgmTgPAAAAAADJBh3nf/Ob38T73//+2LZt2+ve8+STT8bSpUtj/vz5ccUVV8QTTzwx2LcDAAAAAIAxY1Bx/pe//GVce+218dJLL73uPXv37o3W1ta48cYb4+mnn47W1tZoa2uLl19+edCHBQAAAACAsaBwnH/wwQfjlltuiZtuuun/vK+xsTEuu+yyqKysjCVLlsTChQvj/vvvH/RhAQAAAABgLKgs+oL3ve99sXTp0qisrPyDgX737t0xe/bsAdfOP//8eOGFF87qfU6ePBl9fX0xceLEmDBhQtFjAsC4VS6X4+TJk1FZWRkTJ+Z8exl7GwAGx94GgNFjuPd24Th/3nnnndV9x44di+rq6gHXzj333Ojp6Tmr1/f19UV3d3fR4wEA/62+vj6qqqpS3sveBoChsbcBYPQYrr1dOM6frerq6jhx4sSAaydOnIhJkyad1ev7/8vDBRdckPYvKONdqVSKHTt2xNy5c6OiomKkjzMumHk+M89n5vl6e3tj586daZ99F2FvjwTPVj4zz2fm+cw8n709Pni28pl5PjPPZ+b5hntvv2Fxfvbs2fH8888PuLZ79+6YN2/eWb2+//9aV1VV5V8WkpRKpYj4/cw90DnMPJ+Z5zPzkZP5f1O3t/N5tvKZeT4zz2fmI8feHts8W/nMPJ+Z5zPzkTNce/sN+0/zy5Yti66urti0aVP09fXFpk2boqurK5YvX/5GvSUAAAAAAIwKwxrnGxoaYuPGjRERUVdXF/fcc0+sXbs2Fi5cGJ2dnXH33XfH29/+9uF8SwAAAAAAGHWG9GVtdu7cOeDj7du3D/h40aJFsWjRoqG8BQAAAAAAjDl533EGAAAAAACICHEeAAAAAADSifMAAAAAAJBMnAcAAAAAgGTiPAAAAAAAJBPnAQAAAAAgmTgPAAAAAADJxHkAAAAAAEgmzgMAAAAAQDJxHgAAAAAAkonzAAAAAACQTJwHAAAAAIBk4jwAAAAAACQT5wEAAAAAIJk4DwAAAAAAycR5AAAAAABIJs4DAAAAAEAycR4AAAAAAJKJ8wAAAAAAkEycBwAAAACAZOI8AAAAAAAkE+cBAAAAACCZOA8AAAAAAMnEeQAAAAAASCbOAwAAAABAMnEeAAAAAACSifMAAAAAAJBMnAcAAAAAgGTiPAAAAAAAJBPnAQAAAAAgmTgPAAAAAADJxHkAAAAAAEgmzgMAAAAAQDJxHgAAAAAAkonzAAAAAACQTJwHAAAAAIBk4jwAAAAAACQT5wEAAAAAIJk4DwAAAAAAycR5AAAAAABIJs4DAAAAAEAycR4AAAAAAJKJ8wAAAAAAkEycBwAAAACAZOI8AAAAAAAkE+cBAAAAACCZOA8AAAAAAMnEeQAAAAAASCbOAwAAAABAMnEeAAAAAACSifMAAAAAAJBMnAcAAAAAgGTiPAAAAAAAJBPnAQAAAAAgmTgPAAAAAADJxHkAAAAAAEgmzgMAAAAAQDJxHgAAAAAAkonzAAAAAACQTJwHAAAAAIBk4jwAAAAAACQT5wEAAAAAIJk4DwAAAAAAycR5AAAAAABIJs4DAAAAAEAycR4AAAAAAJKJ8wAAAAAAkEycBwAAAACAZOI8AAAAAAAkE+cBAAAAACCZOA8AAAAAAMnEeQAAAAAASCbOAwAAAABAMnEeAAAAAACSifMAAAAAAJBMnAcAAAAAgGTiPAAAAAAAJBPnAQAAAAAgmTgPAAAAAADJxHkAAAAAAEgmzgMAAAAAQDJxHgAAAAAAkonzAAAAAACQTJwHAAAAAIBkheP8oUOHoqWlJRobG6OpqSna29ujr6/vjPd+73vfi0svvTQuvvjiWLp0aTz22GNDPjAAAAAAAIx2heN8W1tb1NTUxJYtW2LDhg2xdevWWLdu3Wn3Pfnkk7F27dr4zne+E88880xcf/310dbWFr/+9a+H49wAAAAAADBqFYrz+/bti66urli5cmVUV1fHjBkzoqWlJdavX3/avS+++GKUy+VTf1VUVMQ555wTlZWVw3Z4AAAAAAAYjQqV8l27dsXUqVNj2rRpp67V1dXFgQMH4siRIzFlypRT16+88sp44IEHYsmSJVFRURETJkyIr371qzF9+vRCByyVSlEqlQq9hsHpn7N55zHzfGaez8zzjeSs7e08nq18Zp7PzPOZeT57e3zwbOUz83xmns/M8w33rAvF+WPHjkV1dfWAa/0f9/T0DIjzr732WsyZMyfa29tjzpw58fDDD8eqVauirq4uLrjggrN+zx07dhQ5IsOgu7t7pI8w7ph5PjPPZ+bjg72dz7OVz8zzmXk+Mx8f7O18nq18Zp7PzPOZ+ehVKM7X1NTE8ePHB1zr/3jSpEkDrt9xxx1x8cUXx4UXXhgRER/60Ifixz/+cTz44IPxuc997qzfc+7cuVFVVVXkmAxSqVSK7u7uqK+vj4qKipE+zrhg5vnMPJ+Z5+vt7R2xP2zb23k8W/nMPJ+Z5zPzfPb2+ODZymfm+cw8n5nnG+69XSjOz5o1Kw4fPhwHDx6M2traiIjYs2dPTJ8+PSZPnjzg3gMHDsS8efMGvlllZZxzzjmFDlhRUeE3VzIzz2fm+cw8n5nnGck5+3XOZ+b5zDyfmecz8zz29vhi5vnMPJ+Z5zPzPMM950LfEHbmzJmxYMGCWLNmTRw9ejT2798fnZ2dsWLFitPuvfTSS+P73/9+PP/883Hy5MnYvHlzbNu2LZYsWTJshwcAAAAAgNGo0GfOR0R0dHTE6tWro7m5OSZOnBhXX311tLS0REREQ0ND3H777bFs2bK4/vrro6KiIlpbW+PVV1+Nt73tbXHPPffEO97xjmH/SQAAAAAAwGhSOM7X1tZGR0fHGX9s+/bt//M3rqyM1tbWaG1tHfzpAAAAAABgDCr0ZW0AAAAAAIChE+cBAAAAACCZOA8AAAAAAMnEeQAAAAAASCbOAwAAAABAMnEeAAAAAACSifMAAAAAAJBMnAcAAAAAgGTiPAAAAAAAJBPnAQAAAAAgmTgPAAAAAADJxHkAAAAAAEgmzgMAAAAAQDJxHgAAAAAAkonzAAAAAACQTJwHAAAAAIBk4jwAAAAAACQT5wEAAAAAIJk4DwAAAAAAycR5AAAAAABIJs4DAAAAAEAycR4AAAAAAJKJ8wAAAAAAkEycBwAAAACAZOI8AAAAAAAkE+cBAAAAACCZOA8AAAAAAMnEeQAAAAAASCbOAwAAAABAMnEeAAAAAACSifMAAAAAAJBMnAcAAAAAgGTiPAAAAAAAJBPnAQAAAAAgmTgPAAAAAADJxHkAAAAAAEgmzgMAAAAAQDJxHgAAAAAAkonzAAAAAACQTJwHAAAAAIBk4jwAAAAAACQT5wEAAAAAIJk4DwAAAAAAycR5AAAAAABIJs4DAAAAAEAycR4AAAAAAJKJ8wAAAAAAkEycBwAAAACAZOI8AAAAAAAkE+cBAAAAACCZOA8AAAAAAMnEeQAAAAAASCbOAwAAAABAMnEeAAAAAACSifMAAAAAAJBMnAcAAAAAgGTiPAAAAAAAJBPnAQAAAAAgmTgPAAAAAADJxHkAAAAAAEgmzgMAAAAAQDJxHgAAAAAAkonzAAAAAACQTJwHAAAAAIBk4jwAAAAAACQT5wEAAAAAIJk4DwAAAAAAycR5AAAAAABIJs4DAAAAAEAycR4AAAAAAJKJ8wAAAAAAkEycBwAAAACAZOI8AAAAAAAkE+cBAAAAACCZOA8AAAAAAMnEeQAAAAAASCbOAwAAAABAMnEeAAAAAACSifMAAAAAAJBMnAcAAAAAgGTiPAAAAAAAJBPnAQAAAAAgmTgPAAAAAADJxHkAAAAAAEgmzgMAAAAAQLLCcf7QoUPR0tISjY2N0dTUFO3t7dHX13fGe7u6uuLDH/5wNDQ0xOLFi2Pt2rVDPjAAAAAAAIx2heN8W1tb1NTUxJYtW2LDhg2xdevWWLdu3Wn37dmzJz71qU/FRz/60XjmmWdi7dq18d3vfjc2b948HOcGAAAAAIBRq1Cc37dvX3R1dcXKlSujuro6ZsyYES0tLbF+/frT7v3BD34Qzc3N8cEPfjAmTJgQc+bMiR/+8IexYMGCYTs8AAAAAACMRpVFbt61a1dMnTo1pk2bdupaXV1dHDhwII4cORJTpkw5df25556L9773vXHzzTfHL37xi3jzm98cH//4x+Paa68tdMBSqRSlUqnQaxic/jmbdx4zz2fm+cw830jO2t7O49nKZ+b5zDyfmeezt8cHz1Y+M89n5vnMPN9wz7pQnD927FhUV1cPuNb/cU9Pz4A4/+qrr8a9994bd911V3zlK1+J7du3x3XXXRdvetOb4gMf+MBZv+eOHTuKHJFh0N3dPdJHGHfMPJ+Z5zPz8cHezufZymfm+cw8n5mPD/Z2Ps9WPjPPZ+b5zHz0KhTna2pq4vjx4wOu9X88adKkAderqqqiubk5LrnkkoiIWLhwYSxfvjweffTRQnF+7ty5UVVVVeSYDFKpVIru7u6or6+PioqKkT7OuGDm+cw8n5nn6+3tHbE/bNvbeTxb+cw8n5nnM/N89vb44NnKZ+b5zDyfmecb7r1dKM7PmjUrDh8+HAcPHoza2tqI+P03fp0+fXpMnjx5wL11dXXR29s74FqpVIpyuVzogBUVFX5zJTPzfGaez8zzmXmekZyzX+d8Zp7PzPOZeT4zz2Nvjy9mns/M85l5PjPPM9xzLvQNYWfOnBkLFiyINWvWxNGjR2P//v3R2dkZK1asOO3ej3zkI/HTn/40fvSjH0W5XI6nnnoqHn744Vi+fPmwHR4AAAAAAEajQnE+IqKjoyP6+vqiubk5rrnmmli0aFG0tLRERERDQ0Ns3LgxIiLe8573RGdnZ9x7772xYMGCuPXWW+Ozn/1sNDc3D+/PAAAAAAAARplCX9YmIqK2tjY6OjrO+GPbt28f8PHixYtj8eLFgzsZAAAAAACMUYU/cx4AAAAAABgacR4AAAAAAJKJ8wAAAAAAkEycBwAAAACAZOI8AAAAAAAkE+cBAAAAACCZOA8AAAAAAMnEeQAAAAAASCbOAwAAAABAMnEeAAAAAACSifMAAAAAAJBMnAcAAAAAgGTiPAAAAAAAJBPnAQAAAAAgmTgPAAAAAADJxHkAAAAAAEgmzgMAAAAAQDJxHgAAAAAAkonzAAAAAACQTJwHAAAAAIBk4jwAAAAAACQT5wEAAAAAIJk4DwAAAAAAycR5AAAAAABIJs4DAAAAAEAycR4AAAAAAJKJ8wAAAAAAkEycBwAAAACAZOI8AAAAAAAkE+cBAAAAACCZOA8AAAAAAMnEeQAAAAAASCbOAwAAAABAMnEeAAAAAACSifMAAAAAAJBMnAcAAAAAgGTiPAAAAAAAJBPnAQAAAAAgmTgPAAAAAADJxHkAAAAAAEgmzgMAAAAAQDJxHgAAAAAAkonzAAAAAACQTJwHAAAAAIBk4jwAAAAAACQT5wEAAAAAIJk4DwAAAAAAycR5AAAAAABIJs4DAAAAAEAycR4AAAAAAJKJ8wAAAAAAkEycBwAAAACAZOI8AAAAAAAkE+cBAAAAACCZOA8AAAAAAMnEeQAAAAAASCbOAwAAAABAMnEeAAAAAACSifMAAAAAAJBMnAcAAAAAgGTiPAAAAAAAJBPnAQAAAAAgmTgPAAAAAADJxHkAAAAAAEgmzgMAAAAAQDJxHgAAAAAAkonzAAAAAACQTJwHAAAAAIBk4jwAAAAAACQT5wEAAAAAIJk4DwAAAAAAycR5AAAAAABIJs4DAAAAAEAycR4AAAAAAJKJ8wAAAAAAkEycBwAAAACAZOI8AAAAAAAkE+cBAAAAACCZOA8AAAAAAMnEeQAAAAAASCbOAwAAAABAMnEeAAAAAACSifMAAAAAAJBMnAcAAAAAgGSF4/yhQ4eipaUlGhsbo6mpKdrb26Ovr+8PvuZXv/pVXHTRRbFt27ZBHxQAAAAAAMaKwnG+ra0tampqYsuWLbFhw4bYunVrrFu37nXvP378eHzmM5+JEydODOWcAAAAAAAwZhSK8/v27Yuurq5YuXJlVFdXx4wZM6KlpSXWr1//uq+5/fbb47LLLhvyQQEAAAAAYKyoLHLzrl27YurUqTFt2rRT1+rq6uLAgQNx5MiRmDJlyoD7H3roodi3b1+0t7dHZ2fnoA5YKpWiVCoN6rUU0z9n885j5vnMPJ+Z5xvJWdvbeTxb+cw8n5nnM/N89vb44NnKZ+b5zDyfmecb7lkXivPHjh2L6urqAdf6P+7p6RkQ5/fs2RN33XVX3HfffVFRUTHoA+7YsWPQr2Vwuru7R/oI446Z5zPzfGY+Ptjb+Txb+cw8n5nnM/Pxwd7O59nKZ+b5zDyfmY9eheJ8TU1NHD9+fMC1/o8nTZp06trvfve7uOmmm+Lzn/98vOUtbxnSAefOnRtVVVVD+ntwdkqlUnR3d0d9ff2Q/oMKZ8/M85l5PjPP19vbO2J/2La383i28pl5PjPPZ+b57O3xwbOVz8zzmXk+M8833Hu7UJyfNWtWHD58OA4ePBi1tbUR8fvPkJ8+fXpMnjz51H3d3d2xd+/eWLVqVaxaterU9U9/+tOxfPnyuO222876PSsqKvzmSmbm+cw8n5nnM/M8Izlnv875zDyfmecz83xmnsfeHl/MPJ+Z5zPzfGaeZ7jnXCjOz5w5MxYsWBBr1qyJ1atXx29/+9vo7OyMFStWDLivsbExnnvuuQHXLrjggvj2t78dTU1NQz81AAAAAACMYhOLvqCjoyP6+vqiubk5rrnmmli0aFG0tLRERERDQ0Ns3Lhx2A8JAAAAAABjSaHPnI+IqK2tjY6OjjP+2Pbt21/3dTt37iz6VgAAAAAAMCYV/sx5AAAAAABgaMR5AAAAAABIJs4DAAAAAEAycR4AAAAAAJKJ8wAAAAAAkEycBwAAAACAZOI8AAAAAAAkE+cBAAAAACCZOA8AAAAAAMnEeQAAAAAASCbOAwAAAABAMnEeAAAAAACSifMAAAAAAJBMnAcAAAAAgGTiPAAAAAAAJBPnAQAAAAAgmTgPAAAAAADJxHkAAAAAAEgmzgMAAAAAQDJxHgAAAAAAkonzAAAAAACQTJwHAAAAAIBk4jwAAAAAACQT5wEAAAAAIJk4DwAAAAAAycR5AAAAAABIJs4DAAAAAEAycR4AAAAAAJKJ8wAAAAAAkEycBwAAAACAZOI8AAAAAAAkE+cBAAAAACCZOA8AAAAAAMnEeQAAAAAASCbOAwAAAABAMnEeAAAAAACSifMAAAAAAJBMnAcAAAAAgGTiPAAAAAAAJBPnAQAAAAAgmTgPAAAAAADJxHkAAAAAAEgmzgMAAAAAQDJxHgAAAAAAkonzAAAAAACQTJwHAAAAAIBk4jwAAAAAACQT5wEAAAAAIJk4DwAAAAAAycR5AAAAAABIJs4DAAAAAEAycR4AAAAAAJKJ8wAAAAAAkEycBwAAAACAZOI8AAAAAAAkE+cBAAAAACCZOA8AAAAAAMnEeQAAAAAASCbOAwAAAABAMnEeAAAAAACSifMAAAAAAJBMnAcAAAAAgGTiPAAAAAAAJBPnAQAAAAAgmTgPAAAAAADJxHkAAAAAAEgmzgMAAAAAQDJxHgAAAAAAkonzAAAAAACQTJwHAAAAAIBk4jwAAAAAACQT5wEAAAAAIJk4DwAAAAAAycR5AAAAAABIJs4DAAAAAEAycR4AAAAAAJKJ8wAAAAAAkEycBwAAAACAZOI8AAAAAAAkE+cBAAAAACCZOA8AAAAAAMnEeQAAAAAASCbOAwAAAABAMnEeAAAAAACSFY7zhw4dipaWlmhsbIympqZob2+Pvr6+M9573333xeWXXx4NDQ1x+eWXx/r164d8YAAAAAAAGO0Kx/m2traoqamJLVu2xIYNG2Lr1q2xbt260+77yU9+Et/4xjfiy1/+cjzzzDNx5513xje/+c147LHHhuPcAAAAAAAwahWK8/v27Yuurq5YuXJlVFdXx4wZM6KlpeWMnxH/8ssvxyc/+cmYP39+TJgwIRoaGqKpqSmeeuqpYTs8AAAAAACMRpVFbt61a1dMnTo1pk2bdupaXV1dHDhwII4cORJTpkw5df1jH/vYgNceOnQonnrqqbj11lsLHbBUKkWpVCr0Gganf87mncfM85l5PjPPN5KztrfzeLbymXk+M89n5vns7fHBs5XPzPOZeT4zzzfcsy4U548dOxbV1dUDrvV/3NPTMyDO/2+vvPJKXHfddTFv3ry46qqrCh1wx44dhe5n6Lq7u0f6COOOmecz83xmPj7Y2/k8W/nMPJ+Z5zPz8cHezufZymfm+cw8n5mPXoXifE1NTRw/fnzAtf6PJ02adMbXPPvss3HjjTdGY2NjfOlLX4rKykJvGXPnzo2qqqpCr2FwSqVSdHd3R319fVRUVIz0ccYFM89n5vnMPF9vb++I/WHb3s7j2cpn5vnMPJ+Z57O3xwfPVj4zz2fm+cw833Dv7UKlfNasWXH48OE4ePBg1NbWRkTEnj17Yvr06TF58uTT7t+wYUN88YtfjBtuuCE+8YlPDOqAFRUVfnMlM/N8Zp7PzPOZeZ6RnLNf53xmns/M85l5PjPPY2+PL2aez8zzmXk+M88z3HMu9A1hZ86cGQsWLIg1a9bE0aNHY//+/dHZ2RkrVqw47d7HHnssbrvttrj77rsHHeYBAAAAAGAsKhTnIyI6Ojqir68vmpub45prrolFixZFS0tLREQ0NDTExo0bIyLiW9/6VpRKpbjhhhuioaHh1F9f+MIXhvdnAAAAAAAAo0yxLwAfEbW1tdHR0XHGH9u+ffup//3www8P/lQAAAAAADCGFf7MeQAAAAAAYGjEeQAAAAAASCbOAwAAAABAMnEeAAAAAACSifMAAAAAAJBMnAcAAAAAgGTiPAAAAAAAJBPnAQAAAAAgmTgPAAAAAADJxHkAAAAAAEgmzgMAAAAAQDJxHgAAAAAAkonzAAAAAACQTJwHAAAAAIBk4jwAAAAAACQT5wEAAAAAIJk4DwAAAAAAycR5AAAAAABIJs4DAAAAAEAycR4AAAAAAJKJ8wAAAAAAkEycBwAAAACAZOI8AAAAAAAkE+cBAAAAACCZOA8AAAAAAMnEeQAAAAAASCbOAwAAAABAMnEeAAAAAACSifMAAAAAAJBMnAcAAAAAgGTiPAAAAAAAJBPnAQAAAAAgmTgPAAAAAADJxHkAAAAAAEgmzgMAAAAAQDJxHgAAAAAAkonzAAAAAACQTJwHAAAAAIBk4jwAAAAAACQT5wEAAAAAIJk4DwAAAAAAycR5AAAAAABIJs4DAAAAAEAycR4AAAAAAJKJ8wAAAAAAkEycBwAAAACAZOI8AAAAAAAkE+cBAAAAACCZOA8AAAAAAMnEeQAAAAAASCbOAwAAAABAMnEeAAAAAACSifMAAAAAAJBMnAcAAAAAgGTiPAAAAAAAJBPnAQAAAAAgmTgPAAAAAADJxHkAAAAAAEgmzgMAAAAAQDJxHgAAAAAAkonzAAAAAACQTJwHAAAAAIBk4jwAAAAAACQT5wEAAAAAIJk4DwAAAAAAycR5AAAAAABIJs4DAAAAAEAycR4AAAAAAJKJ8wAAAAAAkEycBwAAAACAZOI8AAAAAAAkE+cBAAAAACCZOA8AAAAAAMnEeQAAAAAASCbOAwAAAABAMnEeAAAAAACSifMAAAAAAJBMnAcAAAAAgGTiPAAAAAAAJBPnAQAAAAAgmTgPAAAAAADJxHkAAAAAAEgmzgMAAAAAQDJxHgAAAAAAkonzAAAAAACQrHCcP3ToULS0tERjY2M0NTVFe3t79PX1nfHeJ598MpYuXRrz58+PK664Ip544okhHxgAAAAAAEa7wnG+ra0tampqYsuWLbFhw4bYunVrrFu37rT79u7dG62trXHjjTfG008/Ha2trdHW1hYvv/zycJwbAAAAAABGrUJxft++fdHV1RUrV66M6urqmDFjRrS0tMT69etPu/fBBx+MxsbGuOyyy6KysjKWLFkSCxcujPvvv3/YDg8AAAAAAKNRZZGbd+3aFVOnTo1p06adulZXVxcHDhyII0eOxJQpU05d3717d8yePXvA688///x44YUXzuq9yuVyRET09vYWOSJDUCqVIuL3M6+oqBjh04wPZp7PzPOZeb7+3dm/SzPY2/k8W/nMPJ+Z5zPzfPb2+ODZymfm+cw8n5nnG+69XSjOHzt2LKqrqwdc6/+4p6dnQJw/073nnntu9PT0nNV7nTx5MiIidu7cWeSIDIMdO3aM9BHGHTPPZ+b5zDxf/y7NfC97O59nK5+Z5zPzfGaez94eHzxb+cw8n5nnM/N8w7W3C8X5mpqaOH78+IBr/R9PmjRpwPXq6uo4ceLEgGsnTpw47b7XPVhlZdTX18fEiRNjwoQJRY4JAONauVyOkydPRmVloTU/JPY2AAyOvQ0Ao8dw7+1Cf5dZs2bF4cOH4+DBg1FbWxsREXv27Inp06fH5MmTB9w7e/bseP755wdc2717d8ybN++s3mvixIlRVVVV5HgAwAixtwFg9LC3AeCPQ6FvCDtz5sxYsGBBrFmzJo4ePRr79++Pzs7OWLFixWn3Llu2LLq6umLTpk3R19cXmzZtiq6urli+fPmwHR4AAAAAAEajCeWCX73+4MGDsXr16ti2bVtMnDgxrr766rjllluioqIiGhoa4vbbb49ly5ZFRMSWLVvia1/7Wrz00kvx1re+NVauXBmLFy9+Q34iAAAAAAAwWhSO8wAAAAAAwNAU+rI2AAAAAADA0InzAAAAAACQTJwHAAAAAIBk4jwAAAAAACQT5wEAAAAAIJk4DwAAAAAAycR5AAAAAABIJs4DAAAAAEAycR4AAAAAAJKJ8wAAAAAAkEycBwAAAACAZOI8AAAAAAAkG3Sc/81vfhPvf//7Y9u2ba97z5NPPhlLly6N+fPnxxVXXBFPPPHEYN8OAAAAAADGjEHF+V/+8pdx7bXXxksvvfS69+zduzdaW1vjxhtvjKeffjpaW1ujra0tXn755UEfFgAAAAAAxoLCcf7BBx+MW265JW666ab/877Gxsa47LLLorKyMpYsWRILFy6M+++/f9CHBQAAAACAsaCy6Ave9773xdKlS6OysvIPBvrdu3fH7NmzB1w7//zz44UXXjir9zl58mT09fXFxIkTY8KECUWPCQDjVrlcjpMnT0ZlZWVMnJjz7WXsbQAYHHsbAEaP4d7bheP8eeedd1b3HTt2LKqrqwdcO/fcc6Onp+esXt/X1xfd3d1FjwcA/Lf6+vqoqqpKeS97GwCGxt4GgNFjuPZ24Th/tqqrq+PEiRMDrp04cSImTZp0Vq/v/y8PF1xwQdq/oIx3pVIpduzYEXPnzo2KioqRPs64YOb5zDyfmefr7e2NnTt3pn32XYS9PRI8W/nMPJ+Z5zPzfPb2+ODZymfm+cw8n5nnG+69/YbF+dmzZ8fzzz8/4Nru3btj3rx5Z/X6/v9rXVVVlX9ZSFIqlSLi9zP3QOcw83xmns/MR07m/03d3s7n2cpn5vnMPJ+Zjxx7e2zzbOUz83xmns/MR85w7e037D/NL1u2LLq6umLTpk3R19cXmzZtiq6urli+fPkb9ZYAAAAAADAqDGucb2hoiI0bN0ZERF1dXdxzzz2xdu3aWLhwYXR2dsbdd98db3/724fzLQEAAAAAYNQZ0pe12blz54CPt2/fPuDjRYsWxaJFi4byFgAAAAAAMObkfccZAAAAAAAgIsR5AAAAAABIJ84DAAAAAEAycR4AAAAAAJKJ8wAAAAAAkEycBwAAAACAZOI8AAAAAAAkE+cBAAAAACCZOA8AAAAAAMnEeQAAAAAASCbOAwAAAABAMnEeAAAAAACSifMAAAAAAJBMnAcAAAAAgGTiPAAAAAAAJBPnAQAAAAAgmTgPAAAAAADJxHkAAAAAAEgmzgMAAAAAQDJxHgAAAAAAkonzAAAAAACQTJwHAAAAAIBk4jwAAAAAACQT5wEAAAAAIJk4DwAAAAAAycR5AAAAAABIJs4DAAAAAEAycR4AAAAAAJKJ8wAAAAAAkEycBwAAAACAZOI8AAAAAAAkE+cBAAAAACCZOA8AAAAAAMnEeQAAAAAASCbOAwAAAABAMnEeAAAAAACSifMAAAAAAJBMnAcAAAAAgGTiPAAAAAAAJBPnAQAAAAAgmTgPAAAAAADJxHkAAAAAAEgmzgMAAAAAQDJxHgAAAAAAkonzAAAAAACQTJwHAAAAAIBk4jwAAAAAACQT5wEAAAAAIJk4DwAAAAAAycR5AAAAAABIJs4DAAAAAEAycR4AAAAAAJKJ8wAAAAAAkEycBwAAAACAZOI8AAAAAAAkE+cBAAAAACCZOA8AAAAAAMnEeQAAAAAASCbOAwAAAABAMnEeAAAAAACSifMAAAAAAJBMnAcAAAAAgGTiPAAAAAAAJBPnAQAAAAAgmTgPAAAAAADJxHkAAAAAAEgmzgMAAAAAQDJxHgAAAAAAkonzAAAAAACQTJwHAAAAAIBk4jwAAAAAACQT5wEAAAAAIJk4DwAAAAAAycR5AAAAAABIJs4DAAAAAEAycR4AAAAAAJKJ8wAAAAAAkEycBwAAAACAZOI8AAAAAAAkE+cBAAAAACCZOA8AAAAAAMnEeQAAAAAASCbOAwAAAABAMnEeAAAAAACSFY7zhw4dipaWlmhsbIympqZob2+Pvr6+M977ve99Ly699NK4+OKLY+nSpfHYY48N+cAAAAAAADDaFY7zbW1tUVNTE1u2bIkNGzbE1q1bY926dafd9+STT8batWvjO9/5TjzzzDNx/fXXR1tbW/z6178ejnMDAAAAAMCoVSjO79u3L7q6umLlypVRXV0dM2bMiJaWlli/fv1p97744otRLpdP/VVRURHnnHNOVFZWDtvhAQAAAABgNCpUynft2hVTp06NadOmnbpWV1cXBw4ciCNHjsSUKVNOXb/yyivjgQceiCVLlkRFRUVMmDAhvvrVr8b06dMLHbBUKkWpVCr0Gganf87mncfM85l5PjPPN5KztrfzeLbymXk+M89n5vns7fHBs5XPzPOZeT4zzzfcsy4U548dOxbV1dUDrvV/3NPTMyDOv/baazFnzpxob2+POXPmxMMPPxyrVq2Kurq6uOCCC876PXfs2FHkiAyD7u7ukT7CuGPm+cw8n5mPD/Z2Ps9WPjPPZ+b5zHx8sLfzebbymXk+M89n5qNXoThfU1MTx48fH3Ct/+NJkyYNuH7HHXfExRdfHBdeeGFERHzoQx+KH//4x/Hggw/G5z73ubN+z7lz50ZVVVWRYzJIpVIpuru7o76+PioqKkb6OOOCmecz83xmnq+3t3fE/rBtb+fxbOUz83xmns/M89nb44NnK5+Z5zPzfGaeb7j3dqE4P2vWrDh8+HAcPHgwamtrIyJiz549MX369Jg8efKAew8cOBDz5s0b+GaVlXHOOecUOmBFRYXfXMnMPJ+Z5zPzfGaeZyTn7Nc5n5nnM/N8Zp7PzPPY2+OLmecz83xmns/M8wz3nAt9Q9iZM2fGggULYs2aNXH06NHYv39/dHZ2xooVK06799JLL43vf//78fzzz8fJkydj8+bNsW3btliyZMmwHR4AAAAAAEajQp85HxHR0dERq1evjubm5pg4cWJcffXV0dLSEhERDQ0Ncfvtt8eyZcvi+uuvj4qKimhtbY1XX3013va2t8U999wT73jHO4b9JwEAAAAAAKNJ4ThfW1sbHR0dZ/yx7du3/8/fuLIyWltbo7W1dfCnAwAAAACAMajQl7UBAAAAAACGTpwHAAAAAIBk4jwAAAAAACQT5wEAAAAAIJk4DwAAAAAAycR5AAAAAABIJs4DAAAAAEAycR4AAAAAAJKJ8wAAAAAAkEycBwAAAACAZOI8AAAAAAAkE+cBAAAAACCZOA8AAAAAAMnEeQAAAAAASCbOAwAAAABAMnEeAAAAAACSifMAAAAAAJBMnAcAAAAAgGTiPAAAAAAAJBPnAQAAAAAgmTgPAAAAAADJxHkAAAAAAEgmzgMAAAAAQDJxHgAAAAAAkonzAAAAAACQTJwHAAAAAIBk4jwAAAAAACQT5wEAAAAAIJk4DwAAAAAAycR5AAAAAABIJs4DAAAAAEAycR4AAAAAAJKJ8wAAAAAAkEycBwAAAACAZOI8AAAAAAAkE+cBAAAAACCZOA8AAAAAAMnEeQAAAAAASCbOAwAAAABAMnEeAAAAAACSifMAAAAAAJBMnAcAAAAAgGTiPAAAAAAAJBPnAQAAAAAgmTgPAAAAAADJxHkAAAAAAEgmzgMAAAAAQDJxHgAAAAAAkonzAAAAAACQTJwHAAAAAIBk4jwAAAAAACQT5wEAAAAAIJk4DwAAAAAAycR5AAAAAABIJs4DAAAAAEAycR4AAAAAAJKJ8wAAAAAAkEycBwAAAACAZOI8AAAAAAAkE+cBAAAAACCZOA8AAAAAAMnEeQAAAAAASCbOAwAAAABAMnEeAAAAAACSifMAAAAAAJBMnAcAAAAAgGTiPAAAAAAAJBPnAQAAAAAgmTgPAAAAAADJxHkAAAAAAEgmzgMAAAAAQDJxHgAAAAAAkonzAAAAAACQTJwHAAAAAIBk4jwAAAAAACQT5wEAAAAAIJk4DwAAAAAAycR5AAAAAABIJs4DAAAAAEAycR4AAAAAAJKJ8wAAAAAAkEycBwAAAACAZOI8AAAAAAAkE+cBAAAAACBZ4Th/6NChaGlpicbGxmhqaor29vbo6+s7471dXV3x4Q9/OBoaGmLx4sWxdu3aIR8YAAAAAABGu8Jxvq2tLWpqamLLli2xYcOG2Lp1a6xbt+60+/bs2ROf+tSn4qMf/Wg888wzsXbt2vjud78bmzdvHo5zAwAAAADAqFUozu/bty+6urpi5cqVUV1dHTNmzIiWlpZYv379aff+4Ac/iObm5vjgBz8YEyZMiDlz5sQPf/jDWLBgwbAdHgAAAAAARqPKIjfv2rUrpk6dGtOmTTt1ra6uLg4cOBBHjhyJKVOmnLr+3HPPxXvf+964+eab4xe/+EW8+c1vjo9//ONx7bXXFjpgqVSKUqlU6DUMTv+czTuPmecz83xmnm8kZ21v5/Fs5TPzfGaez8zz2dvjg2crn5nnM/N8Zp5vuGddKM4fO3YsqqurB1zr/7inp2dAnH/11Vfj3nvvjbvuuiu+8pWvxPbt2+O6666LN73pTfGBD3zgrN9zx44dRY7IMOju7h7pI4w7Zp7PzPOZ+fhgb+fzbOUz83xmns/Mxwd7O59nK5+Z5zPzfGY+ehWK8zU1NXH8+PEB1/o/njRp0oDrVVVV0dzcHJdccklERCxcuDCWL18ejz76aKE4P3fu3KiqqipyTAapVCpFd3d31NfXR0VFxUgfZ1ww83xmns/M8/X29o7YH7bt7TyerXxmns/M85l5Pnt7fPBs5TPzfGaez8zzDffeLhTnZ82aFYcPH46DBw9GbW1tRPz+G79Onz49Jk+ePODeurq66O3tHXCtVCpFuVwudMCKigq/uZKZeT4zz2fm+cw8z0jO2a9zPjPPZ+b5zDyfmeext8cXM89n5vnMPJ+Z5xnuORf6hrAzZ86MBQsWxJo1a+Lo0aOxf//+6OzsjBUrVpx270c+8pH46U9/Gj/60Y+iXC7HU089FQ8//HAsX7582A4PAAAAAACjUaE4HxHR0dERfX190dzcHNdcc00sWrQoWlpaIiKioaEhNm7cGBER73nPe6KzszPuvffeWLBgQdx6663x2c9+Npqbm4f3ZwAAAAAAAKNMoS9rExFRW1sbHR0dZ/yx7du3D/h48eLFsXjx4sGdDAAAAAAAxqjCnzkPAAAAAAAMjTgPAAAAAADJxHkAAAAAAEgmzgMAAAAAQDJxHgAAAAAAkonzAAAAAACQTJwHAAAAAIBk4jwAAAAAACQT5wEAAAAAIJk4DwAAAAAAycR5AAAAAABIJs4DAAAAAEAycR4AAAAAAJKJ8wAAAAAAkEycBwAAAACAZOI8AAAAAAAkE+cBAAAAACCZOA8AAAAAAMnEeQAAAAAASCbOAwAAAABAMnEeAAAAAACSifMAAAAAAJBMnAcAAAAAgGTiPAAAAAAAJBPnAQAAAAAgmTgPAAAAAADJxHkAAAAAAEgmzgMAAAAAQDJxHgAAAAAAkonzAAAAAACQTJwHAAAAAIBk4jwAAAAAACQT5wEAAAAAIJk4DwAAAAAAycR5AAAAAABIJs4DAAAAAEAycR4AAAAAAJKJ8wAAAAAAkEycBwAAAACAZOI8AAAAAAAkE+cBAAAAACCZOA8AAAAAAMnEeQAAAAAASCbOAwAAAABAMnEeAAAAAACSifMAAAAAAJBMnAcAAAAAgGTiPAAAAAAAJBPnAQAAAAAgmTgPAAAAAADJxHkAAAAAAEgmzgMAAAAAQDJxHgAAAAAAkonzAAAAAACQTJwHAAAAAIBk4jwAAAAAACQT5wEAAAAAIJk4DwAAAAAAycR5AAAAAABIJs4DAAAAAEAycR4AAAAAAJKJ8wAAAAAAkEycBwAAAACAZOI8AAAAAAAkE+cBAAAAACCZOA8AAAAAAMnEeQAAAAAASCbOAwAAAABAMnEeAAAAAACSifMAAAAAAJBMnAcAAAAAgGTiPAAAAAAAJBPnAQAAAAAgmTgPAAAAAADJxHkAAAAAAEgmzgMAAAAAQDJxHgAAAAAAkonzAAAAAACQTJwHAAAAAIBk4jwAAAAAACQT5wEAAAAAIJk4DwAAAAAAycR5AAAAAABIJs4DAAAAAEAycR4AAAAAAJIVjvOHDh2KlpaWaGxsjKampmhvb4++vr4/+Jpf/epXcdFFF8W2bdsGfVAAAAAAABgrCsf5tra2qKmpiS1btsSGDRti69atsW7dute9//jx4/GZz3wmTpw4MZRzAgAAAADAmFEozu/bty+6urpi5cqVUV1dHTNmzIiWlpZYv379677m9ttvj8suu2zIBwUAAAAAgLGissjNu3btiqlTp8a0adNOXaurq4sDBw7EkSNHYsqUKQPuf+ihh2Lfvn3R3t4enZ2dgzpgqVSKUqk0qNdSTP+czTuPmecz83xmnm8kZ21v5/Fs5TPzfGaez8zz2dvjg2crn5nnM/N8Zp5vuGddKM4fO3YsqqurB1zr/7inp2dAnN+zZ0/cddddcd9990VFRcWgD7hjx45Bv5bB6e7uHukjjDtmns/M85n5+GBv5/Ns5TPzfGaez8zHB3s7n2crn5nnM/N8Zj56FYrzNTU1cfz48QHX+j+eNGnSqWu/+93v4qabborPf/7z8Za3vGVIB5w7d25UVVUN6e/B2SmVStHd3R319fVD+g8qnD0zz2fm+cw8X29v74j9YdvezuPZymfm+cw8n5nns7fHB89WPjPPZ+b5zDzfcO/tQnF+1qxZcfjw4Th48GDU1tZGxO8/Q3769OkxefLkU/d1d3fH3r17Y9WqVbFq1apT1z/96U/H8uXL47bbbjvr96yoqPCbK5mZ5zPzfGaez8zzjOSc/TrnM/N8Zp7PzPOZeR57e3wx83xmns/M85l5nuGec6E4P3PmzFiwYEGsWbMmVq9eHb/97W+js7MzVqxYMeC+xsbGeO655wZcu+CCC+Lb3/52NDU1Df3UAAAAAAAwik0s+oKOjo7o6+uL5ubmuOaaa2LRokXR0tISERENDQ2xcePGYT8kAAAAAACMJYU+cz4iora2Njo6Os74Y9u3b3/d1+3cubPoWwEAAAAAwJhU+DPnAQAAAACAoRHnAQAAAAAgmTgPAAAAAADJxHkAAAAAAEgmzgMAAAAAQDJxHgAAAAAAkonzAAAAAACQTJwHAAAAAIBk4jwAAAAAACQT5wEAAAAAIJk4DwAAAAAAycR5AAAAAABIJs4DAAAAAEAycR4AAAAAAJKJ8wAAAAAAkEycBwAAAACAZOI8AAAAAAAkE+cBAAAAACCZOA8AAAAAAMnEeQAAAAAASCbOAwAAAABAMnEeAAAAAACSifMAAAAAAJBMnAcAAAAAgGTiPAAAAAAAJBPnAQAAAAAgmTgPAAAAAADJxHkAAAAAAEgmzgMAAAAAQDJxHgAAAAAAkonzAAAAAACQTJwHAAAAAIBk4jwAAAAAACQT5wEAAAAAIJk4DwAAAAAAycR5AAAAAABIJs4DAAAAAEAycR4AAAAAAJKJ8wAAAAAAkEycBwAAAACAZOI8AAAAAAAkE+cBAAAAACCZOA8AAAAAAMnEeQAAAAAASCbOAwAAAABAMnEeAAAAAACSifMAAAAAAJBMnAcAAAAAgGTiPAAAAAAAJBPnAQAAAAAgmTgPAAAAAADJxHkAAAAAAEgmzgMAAAAAQDJxHgAAAAAAkonzAAAAAACQTJwHAAAAAIBk4jwAAAAAACQT5wEAAAAAIJk4DwAAAAAAycR5AAAAAABIJs4DAAAAAEAycR4AAAAAAJKJ8wAAAAAAkEycBwAAAACAZOI8AAAAAAAkE+cBAAAAACCZOA8AAAAAAMnEeQAAAAAASCbOAwAAAABAMnEeAAAAAACSifMAAAAAAJBMnAcAAAAAgGTiPAAAAAAAJBPnAQAAAAAgmTgPAAAAAADJxHkAAAAAAEgmzgMAAAAAQDJxHgAAAAAAkonzAAAAAACQTJwHAAAAAIBk4jwAAAAAACQT5wEAAAAAIJk4DwAAAAAAycR5AAAAAABIVjjOHzp0KFpaWqKxsTGampqivb09+vr6znjvfffdF5dffnk0NDTE5ZdfHuvXrx/ygQEAAAAAYLQrHOfb2tqipqYmtmzZEhs2bIitW7fGunXrTrvvJz/5SXzjG9+IL3/5y/HMM8/EnXfeGd/85jfjscceG45zAwAAAADAqFUozu/bty+6urpi5cqVUV1dHTNmzIiWlpYzfkb8yy+/HJ/85Cdj/vz5MWHChGhoaIimpqZ46qmnhu3wAAAAAAAwGlUWuXnXrl0xderUmDZt2qlrdXV1ceDAgThy5EhMmTLl1PWPfexjA1576NCheOqpp+LWW28tdMBSqRSlUqnQaxic/jmbdx4zz2fm+cw830jO2t7O49nKZ+b5zDyfmeezt8cHz1Y+M89n5vnMPN9wz7pQnD927FhUV1cPuNb/cU9Pz4A4/7+98sorcd1118W8efPiqquuKnTAHTt2FLqfoevu7h7pI4w7Zp7PzPOZ+fhgb+fzbOUz83xmns/Mxwd7O59nK5+Z5zPzfGY+ehWK8zU1NXH8+PEB1/o/njRp0hlf8+yzz8aNN94YjY2N8aUvfSkqKwu9ZcydOzeqqqoKvYbBKZVK0d3dHfX19VFRUTHSxxkXzDyfmecz83y9vb0j9odtezuPZyufmecz83xmns/eHh88W/nMPJ+Z5zPzfMO9twuV8lmzZsXhw4fj4MGDUVtbGxERe/bsienTp8fkyZNPu3/Dhg3xxS9+MW644Yb4xCc+MagDVlRU+M2VzMzzmXk+M89n5nlGcs5+nfOZeT4zz2fm+cw8j709vph5PjPPZ+b5zDzPcM+50DeEnTlzZixYsCDWrFkTR48ejf3790dnZ2esWLHitHsfe+yxuO222+Luu+8edJgHAAAAAICxqFCcj4jo6OiIvr6+aG5ujmuuuSYWLVoULS0tERHR0NAQGzdujIiIb33rW1EqleKGG26IhoaGU3994QtfGN6fAQAAAAAAjDLFvgB8RNTW1kZHR8cZf2z79u2n/vfDDz88+FMBAAAAAMAYVvgz5wEAAAAAgKER5wEAAAAAIJk4DwAAAAAAycR5AAAAAABIJs4DAAAAAEAycR4AAAAAAJKJ8wAAAAAAkEycBwAAAACAZOI8AAAAAAAkE+cBAAAAACCZOA8AAAAAAMnEeQAAAAAASCbOAwAAAABAMnEeAAAAAACSifMAAAAAAJBMnAcAAAAAgGTiPAAAAAAAJBPnAQAAAAAgmTgPAAAAAADJxHkAAAAAAEgmzgMAAAAAQDJxHgAAAAAAkonzAAAAAACQTJwHAAAAAIBk4jwAAAAAACQT5wEAAAAAIJk4DwAAAAAAycR5AAAAAABIJs4DAAAAAEAycR4AAAAAAJKJ8wAAAAAAkEycBwAAAACAZOI8AAAAAAAkE+cBAAAAACCZOA8AAAAAAMnEeQAAAAAASCbOAwAAAABAMnEeAAAAAACSifMAAAAAAJBMnAcAAAAAgGTiPAAAAAAAJBPnAQAAAAAgmTgPAAAAAADJxHkAAAAAAEgmzgMAAAAAQDJxHgAAAAAAkonzAAAAAACQTJwHAAAAAIBk4jwAAAAAACQT5wEAAAAAIJk4DwAAAAAAycR5AAAAAABIJs4DAAAAAEAycR4AAAAAAJKJ8wAAAAAAkEycBwAAAACAZOI8AAAAAAAkE+cBAAAAACCZOA8AAAAAAMnEeQAAAAAASCbOAwAAAABAMnEeAAAAAACSifMAAAAAAJBMnAcAAAAAgGTiPAAAAAAAJBPnAQAAAAAgmTgPAAAAAADJxHkAAAAAAEgmzgMAAAAAQDJxHgAAAAAAkonzAAAAAACQTJwHAAAAAIBk4jwAAAAAACQT5wEAAAAAIJk4DwAAAAAAycR5AAAAAABIJs4DAAAAAEAycR4AAAAAAJKJ8wAAAAAAkEycBwAAAACAZOI8AAAAAAAkE+cBAAAAACCZOA8AAAAAAMnEeQAAAAAASFY4zh86dChaWlqisbExmpqaor29Pfr6+s5475NPPhlLly6N+fPnxxVXXBFPPPHEkA8MAAAAAACjXeE439bWFjU1NbFly5bYsGFDbN26NdatW3fafXv37o3W1ta48cYb4+mnn47W1tZoa2uLl19+eTjODQAAAAAAo1ahOL9v377o6uqKlStXRnV1dcyYMSNaWlpi/fr1p9374IMPRmNjY1x22WVRWVkZS5YsiYULF8b9998/bIcHAAAAAIDRqLLIzbt27YqpU6fGtGnTTl2rq6uLAwcOxJEjR2LKlCmnru/evTtmz5494PXnn39+vPDCC2f1XuVyOSIient7ixyRISiVShHx+5lXVFSM8GnGBzPPZ+b5zDxf/+7s36UZ7O18nq18Zp7PzPOZeT57e3zwbOUz83xmns/M8w333i4U548dOxbV1dUDrvV/3NPTMyDOn+nec889N3p6es7qvU6ePBkRETt37ixyRIbBjh07RvoI446Z5zPzfGaer3+XZr6XvZ3Ps5XPzPOZeT4zz2dvjw+erXxmns/M85l5vuHa24XifE1NTRw/fnzAtf6PJ02aNOB6dXV1nDhxYsC1EydOnHbf6x6ssjLq6+tj4sSJMWHChCLHBIBxrVwux8mTJ6OystCaHxJ7GwAGx94GgNFjuPd2ob/LrFmz4vDhw3Hw4MGora2NiIg9e/bE9OnTY/LkyQPunT17djz//PMDru3evTvmzZt3Vu81ceLEqKqqKnI8AGCE2NsAMHrY2wDwx6HQN4SdOXNmLFiwINasWRNHjx6N/fv3R2dnZ6xYseK0e5ctWxZdXV2xadOm6Ovri02bNkVXV1csX7582A4PAAAAAACj0YRywa9ef/DgwVi9enVs27YtJk6cGFdffXXccsstUVFREQ0NDXH77bfHsmXLIiJiy5Yt8bWvfS1eeumleOtb3xorV66MxYsXvyE/EQAAAAAAGC0Kx3kAAAAAAGBoCn1ZGwAAAAAAYOjEeQAAAAAASCbOAwAAAABAMnEeAAAAAACSifMAAAAAAJBsROP8oUOHoqWlJRobG6OpqSna29ujr6/vjPc++eSTsXTp0pg/f35cccUV8cQTTySfdmwoMvP77rsvLr/88mhoaIjLL7881q9fn3zasaHIzPv96le/iosuuii2bduWdMqxpcjMu7q64sMf/nA0NDTE4sWLY+3atcmnHRuKzPx73/teXHrppXHxxRfH0qVL47HHHks+7djym9/8Jt7//vf/wX9eDNcOtbfz2dv57O189nY+e3vk2Ntjm72dz97OZ2/ns7dHTsreLo+gv/iLvyh/5jOfKff09JRfeuml8pVXXln+h3/4h9Pu+7d/+7dyfX19+fHHHy+/9tpr5UceeaR84YUXlv/jP/5jBE49up3tzB9//PFyY2Njefv27eWTJ0+Wn3nmmXJjY2N58+bNI3Dq0e1sZ96vp6enfNVVV5Vnz55d/ud//ufEk44dZzvz3bt3ly+66KLyAw88UD558mT5X//1X8vvete7yo8++ugInHp0O9uZ/+xnPyu/5z3vKe/Zs6dcLpfLmzdvLs+ZM6e8f//+7COPCU8//XT5sssu+4P/vBjOHWpv57O389nb+eztfPb2yLC3xz57O5+9nc/ezmdvj4ysvT1icX7v3r3l2bNnDzjsI488Ur7kkktOu/cb3/hG+a/+6q8GXPvrv/7r8t/93d+94eccS4rM/Pvf/3557dq1A679zd/8TfmOO+54w885lhSZeb/Pfvaz5W9+85v+ZWGQisx89erV5ZtvvnnAtRdffLH8n//5n2/4OceSIjP/7ne/W373u99d3r17d/nkyZPlxx9/vFxfX1/+93//98wjjwkPPPBA+ZJLLik/8sgjf/CfF8O1Q+3tfPZ2Pns7n72dz94eGfb22Gdv57O389nb+eztkZG5t0fsy9rs2rUrpk6dGtOmTTt1ra6uLg4cOBBHjhwZcO/u3btj9uzZA66df/758cILL6ScdawoMvOPfexj8alPferUx4cOHYqnnnoq5s2bl3besaDIzCMiHnroodi3b19cf/31mcccU4rM/Lnnnos/+ZM/iZtvvjmampriiiuuiK6urjjvvPOyjz2qFZn5lVdeGbW1tbFkyZJ45zvfGTfeeGPceeedMX369Oxjj3rve9/74vHHH48lS5b8wfuGa4fa2/ns7Xz2dj57O5+9PTLs7bHP3s5nb+ezt/PZ2yMjc2+PWJw/duxYVFdXD7jW/3FPT8//ee+555572n38YUVm/r+98sor8clPfjLmzZsXV1111Rt6xrGmyMz37NkTd911V3z961+PioqKtDOONUVm/uqrr8a9994by5Yti1/84hexevXq+PKXvxybN29OO+9YUGTmr732WsyZMyf+8R//MZ599tlYvXp1rFq1Knbu3Jl23rHivPPOi8rKyv/zvuHaofZ2Pns7n72dz97OZ2+PDHt77LO389nb+eztfPb2yMjc2yMW52tqauL48eMDrvV/PGnSpAHXq6ur48SJEwOunThx4rT7+MOKzLzfs88+GytWrIi3v/3t8fd///dn9RuT/3G2M//d734XN910U3z+85+Pt7zlLalnHGuK/D6vqqqK5ubmuOSSS6KysjIWLlwYy5cvj0cffTTtvGNBkZnfcccdMWvWrLjwwgujqqoqPvShD8X8+fPjwQcfTDvveDNcO9Tezmdv57O389nb+eztP2729uhlb+ezt/PZ2/ns7T9uw7FDRyzOz5o1Kw4fPhwHDx48dW3Pnj0xffr0mDx58oB7Z8+eHbt27Rpwbffu3TFr1qyUs44VRWYeEbFhw4b4+Mc/Hn/5l38ZX//616OqqirzuGPC2c68u7s79u7dG6tWrYrGxsZobGyMiIhPf/rTcdttt2Ufe1Qr8vu8rq4uent7B1wrlUpRLpdTzjpWFJn5gQMHTpt5ZWVlnHPOOSlnHY+Ga4fa2/ns7Xz2dj57O5+9/cfN3h697O189nY+ezufvf3HbVh26OC/NP7Q/fmf/3n5pptuKv/Xf/3Xqe823NHRcdp9u3fvLtfX15cfeeSRU9/5tr6+vvziiy+OwKlHt7Od+ebNm8vvfOc7yz//+c9H4JRjy9nO/P/nG9QM3tnO/J/+6Z/Kc+fOLT/00EPlkydPlru6usrz588v/+QnPxmBU49uZzvzu+66q9zU1FT+l3/5l3KpVCo/+uij5fr6+vKOHTtG4NRjxx/658Vw7lB7O5+9nc/ezmdv57O3R5a9PXbZ2/ns7Xz2dj57e2S90Xt7ROP8K6+8Um5tbS2/613vKr/73e8u33nnneW+vr5yuVwuz58/v/yjH/3o1L0///nPy8uWLSvPnz+/fOWVV5Z/9rOfjdSxR7WznflVV11VnjNnTnn+/PkD/vrbv/3bkTz+qFTk9/n/5l8WBq/IzH/2s5+V/+zP/qzc0NBQbm5uLt93330jdexR7Wxn/tprr5U7OjrKf/qnf1q++OKLyx/84Af9oWQY/P//vHijdqi9nc/ezmdv57O389nbI8veHrvs7Xz2dj57O5+9PbLe6L09oVz2/ycBAAAAAIBMI/Y15wEAAAAAYLwS5wEAAAAAIJk4DwAAAAAAycR5AAAAAABIJs4DAAAAAEAycR4AAAAAAJKJ8wAAAAAAkEycBwAAAACAZOI8AAAAAAAkE+cBAAAAACCZOA8AAAAAAMn+H9634WOKQrI5AAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 1500x1000 with 6 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "_, ax = plt.subplots(2, 3, figsize=(15, 10), sharex=True, sharey=True,\n",
    "                     constrained_layout=True)\n",
    "\n",
    "trace_hmm = trace_hm.posterior.stack(samples={\"chain\", \"draw\"})\n",
    "trace_upp = trace_up.posterior.stack(samples={\"chain\", \"draw\"})\n",
    "\n",
    "\n",
    "ax = np.ravel(ax)\n",
    "j, k = 0, N\n",
    "x_range = np.linspace(x_m.min(), x_m.max(), 10)\n",
    "\n",
    "for i in range(M):\n",
    "    ax[i].scatter(x_m[j:k], y_m[j:k])\n",
    "    ax[i].set_xlabel(f'x_{i}')\n",
    "    ax[i].set_ylabel(f'y_{i}', labelpad=17, rotation=0)\n",
    "    alpha_m = trace_hmm['α'].sel({'α_dim_0':i}).mean().item()\n",
    "    beta_m  = trace_hmm['β'].sel({'β_dim_0':i}).mean().item()\n",
    "    ax[i].plot(x_range, alpha_m + beta_m * x_range, c='r',\n",
    "               label=f'y = {alpha_m:.2f} + {beta_m:.2f} * x'+' - иерахич. модель')\n",
    "\n",
    "    # alpha_o = trace_upp['α'].sel({'α_dim_0':i}).mean().item()\n",
    "    # beta_o = trace_upp['β'].sel({'β_dim_0':i}).mean().item()\n",
    "    # ax[i].plot(x_range, alpha_o + beta_o * x_range, 'g--',\n",
    "    #            label=f'y = {alpha_o:.2f} + {beta_o:.2f} * x'+' - неиерахич. модель')\n",
    "\n",
    "    ax[i].legend()\n",
    "    # plt.xlim(x_m.min()-1, x_m.max()+1)\n",
    "    # plt.ylim(y_m.min()-1, y_m.max()+1)\n",
    "    j += N\n",
    "    k += N\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "plmZGZU7Baxh"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "QxETrzWcBazm"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "jeTdTvnbBa2O"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "pejFCqQrBa4m"
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "colab": {
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
